{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb66c91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from isatools.model import (\n",
    "    Comment,\n",
    "    Investigation,\n",
    "    Study,\n",
    "    StudyFactor,\n",
    "    FactorValue,\n",
    "    OntologyAnnotation,\n",
    "    Characteristic,\n",
    "    OntologySource,\n",
    "    Material,\n",
    "    Sample,\n",
    "    Source,\n",
    "    Protocol,\n",
    "    ProtocolParameter,\n",
    "    ProtocolComponent,\n",
    "    ParameterValue,\n",
    "    Process,\n",
    "    Publication,\n",
    "    Person,\n",
    "    Assay,\n",
    "    DataFile,\n",
    "    plink\n",
    ")\n",
    "import datetime\n",
    "import os\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "511c7560",
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation = Investigation()\n",
    "# i_comment = Comment(name=\"i_comment\", value=\"i_value\")\n",
    "# investigation.comments.append(i_comment)\n",
    "\n",
    "\n",
    "# Declaring the Ontologies and Vocabularies used in the ISA Study\n",
    "# dummy_onto=OntologySource(name=\"Dumbo\",description=\"\")\n",
    "chebi=OntologySource(name=\"CHEBI\",description=\"Chemical Entity of Biological Interest\")\n",
    "efo=OntologySource(name=\"EFO\", description=\"Experimental Factor Ontology\")\n",
    "obi = OntologySource(name='OBI', description=\"Ontology for Biomedical Investigations\")\n",
    "pato = OntologySource(name='PATO', description=\"Phenotype and Trait Ontology\")\n",
    "ncbitaxon = OntologySource(name=\"NCIBTaxon\", description=\"NCBI Taxonomy\")\n",
    "investigation.ontology_source_references=[chebi,efo,obi,pato,ncbitaxon]\n",
    "\n",
    "\n",
    "study = Study(filename=\"s_BII-S-9-synthesic.txt\")\n",
    "# st_comment = Comment(name=\"st_comment\", value=\"st_value\")\n",
    "# study.comments.append(st_comment)\n",
    "study.identifier = \"BII-S-9-synth\"\n",
    "study.title = \"cross-omics synthetic experiment\"\n",
    "study.description = \"cross-omics experiment testing ISA-API support for sample aliquoting at study or assay level\"\n",
    "study.submission_date = \"15/08/2021\"\n",
    "study.public_release_date = \"15/08/2021\"\n",
    "\n",
    "# These NCBI SRA related ISA Comments fields are required and must be present for the ISA SRAconverter is to be invoked later\n",
    "src_comment_sra1 = Comment(name=\"SRA Broker Name\", value=\"OXFORD\")\n",
    "src_comment_sra2 = Comment(name=\"SRA Center Name\", value=\"OXFORD\")\n",
    "src_comment_sra3 = Comment(name=\"SRA Center Project Name\", value=\"OXFORD\")\n",
    "src_comment_sra4 = Comment(name=\"SRA Lab Name\", value=\"Oxford e-Research Centre\")\n",
    "src_comment_sra5 = Comment(name=\"SRA Submission Action\", value=\"ADD\")\n",
    "study.comments.append(src_comment_sra1)\n",
    "study.comments.append(src_comment_sra2)\n",
    "study.comments.append(src_comment_sra3)\n",
    "study.comments.append(src_comment_sra4)\n",
    "study.comments.append(src_comment_sra5)\n",
    "\n",
    "# These ISA Comments are optional and may be used to report funding information\n",
    "src_comment_st1 = Comment(name=\"Study Funding Agency\", value=\"\")\n",
    "src_comment_st2 = Comment(name=\"Study Grant Number\", value=\"\")\n",
    "study.comments.append(src_comment_st1)\n",
    "study.comments.append(src_comment_st2)\n",
    "   \n",
    "# Declaring all the protocols used in the ISA study. Note also the declaration of Protocol Parameters when needed.\n",
    "study.protocols = [\n",
    "    Protocol(name=\"environmental material collection - standard procedure 1\",\n",
    "             description=\"Waters samples were prefiltered through a 1.6 um GF/A glass fibre filter to reduce Eukaryotic contamination. Filtrate was then collected on a 0.2 um Sterivex (millipore) filter which was frozen in liquid nitrogen until nucelic acid extraction. CO2 bubbled through 11000 L mesocosm to simulate ocean acidification predicted conditions. Then phosphate and nitrate were added to induce a phytoplankton bloom.\",\n",
    "             protocol_type=OntologyAnnotation(term=\"sample collection\"),\n",
    "             parameters=[\n",
    "              ProtocolParameter(parameter_name=OntologyAnnotation(term=\"filter pore size\"))\n",
    "             ]\n",
    "            ),\n",
    "    Protocol(\n",
    "        name=\"nucleic acid extraction\",\n",
    "        description=\"Total nucleic acid extraction was done as quickly as possible using the method of Neufeld et al, 2007.\",\n",
    "        protocol_type=OntologyAnnotation(term=\"nucleic acid extraction\")\n",
    "    ),\n",
    "    Protocol(\n",
    "        name=\"sample aliquoting - standard procedure 3\",\n",
    "        description=\"splitting collected samples into aliquots\",\n",
    "        protocol_type=OntologyAnnotation(term=\"aliquoting\") \n",
    "    ),\n",
    "    Protocol(\n",
    "        name=\"genomic DNA extraction - standard procedure 4\",\n",
    "        description=\"superscript+random hexamer primer\",\n",
    "        protocol_type=OntologyAnnotation(term=\"nucleic acid extraction\")\n",
    "    ),\n",
    "    Protocol(\n",
    "        name=\"reverse transcription - standard procedure 5\",\n",
    "        description=\"\",\n",
    "        protocol_type=OntologyAnnotation(term=\"reverse transcription\"),\n",
    "    ),\n",
    "     Protocol(\n",
    "        name=\"library construction\",\n",
    "        description=\"\",\n",
    "        protocol_type=OntologyAnnotation(term=\"library construction\"),\n",
    "        parameters=[\n",
    "            ProtocolParameter(parameter_name=OntologyAnnotation(term=\"library strategy\")),\n",
    "            ProtocolParameter(parameter_name=OntologyAnnotation(term=\"library layout\")),\n",
    "            ProtocolParameter(parameter_name=OntologyAnnotation(term=\"library selection\"))\n",
    "        ]\n",
    "    ),   \n",
    "\n",
    "    Protocol(\n",
    "        name=\"nucleic acid sequencing\", #pyrosequencing - standard procedure 6\",\n",
    "        description=\"1. Sample Input and Fragmentation: The Genome Sequencer FLX System supports the sequencing of samples from a wide variety of starting materials including genomic DNA, PCR products, BACs, and cDNA. Samples such as genomic DNA and BACs are fractionated into small, 300- to 800-base pair fragments. For smaller samples, such as small non-coding RNA or PCR amplicons, fragmentation is not required. Instead, short PCR products amplified using Genome Sequencer fusion primers can be used for immobilization onto DNA capture beads as shown below.\",\n",
    "        protocol_type=OntologyAnnotation(term=\"nucleic acid sequencing\"),\n",
    "        parameters=[\n",
    "            ProtocolParameter(parameter_name=OntologyAnnotation(term=\"sequencing instrument\"))\n",
    "        ]\n",
    "    ),\n",
    "    Protocol(\n",
    "        name=\"sequence analysis - standard procedure 7\",\n",
    "        description=\"\",\n",
    "        protocol_type=OntologyAnnotation(term=\"data transformation\")\n",
    "    )\n",
    "]\n",
    "\n",
    "\n",
    "# Adding a Study Design descriptor to the ISA Study object\n",
    "intervention_design = OntologyAnnotation(term_source=obi)\n",
    "intervention_design.term = \"intervention design\"\n",
    "intervention_design.term_accession = \"http://purl.obolibrary.org/obo/OBI_0000115\"\n",
    "study.design_descriptors.append(intervention_design)\n",
    "\n",
    "\n",
    "# Declaring the Study Factors\n",
    "study.factors = [\n",
    "    StudyFactor(name=\"compound\",factor_type=OntologyAnnotation(term=\"chemical substance\",\n",
    "                                                              term_accession=\"http://purl.obolibrary.org/obo/CHEBI_59999\",\n",
    "                                                              term_source=chebi)),\n",
    "    StudyFactor(name=\"dose\",factor_type=OntologyAnnotation(term=\"dose\", term_accession=\"http://www.ebi.ac.uk/efo/EFO_0000428\",term_source=efo)),\n",
    "    StudyFactor(name=\"collection time\",factor_type=OntologyAnnotation(term=\"time\", term_accession=\"http://purl.obolibrary.org/obo/PATO_0000165\", term_source=pato))\n",
    "]\n",
    "\n",
    "# Associating the levels to each of the Study Factor.\n",
    "fv1 = FactorValue(factor_name=study.factors[0], value=OntologyAnnotation(term=\"atorvastatin\"))\n",
    "fv2 = FactorValue(factor_name=study.factors[1], value=OntologyAnnotation(term=\"high dose\"))\n",
    "fv3 = FactorValue(factor_name=study.factors[1], value=OntologyAnnotation(term=\"low dose\"))\n",
    "fv4 = FactorValue(factor_name=study.factors[2], value=\"2 months\")\n",
    "fv5 = FactorValue(factor_name=study.factors[2], value=\"18 months\")\n",
    "\n",
    "\n",
    "# Adding the publications associated to the study\n",
    "study.publications = [\n",
    "    Publication(doi=\"10.1371/journal.pone.0003042\",pubmed_id=\"18725995\",\n",
    "                title=\"Detection of large numbers of novel sequences in the metatranscriptomes of complex marine microbial communities.\",\n",
    "                status=OntologyAnnotation(term=\"indexed in PubMed\"),\n",
    "                author_list=\"Gilbert JA, Field D, Huang Y, Edwards R, Li W, Gilna P, Joint I.\")  \n",
    "]\n",
    "\n",
    "\n",
    "# Adding the authors of the study\n",
    "study.contacts = [\n",
    "    Person(first_name=\"Rex\", last_name=\"Durand\", affiliation=\"LHC Laboratory\", email=\"rex.durand@lhcl.ac.uk\",\n",
    "           address=\"Nevsky perspective, Bournemouth, United Kingdom\",\n",
    "           comments=[Comment(name=\"Study Person REF\", value=\"\")],\n",
    "            roles=[OntologyAnnotation(term=\"principal investigator role\"),\n",
    "                   OntologyAnnotation(term=\"SRA Inform On Status\"),\n",
    "                   OntologyAnnotation(term=\"SRA Inform On Error\")]\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1875f05e",
   "metadata": {},
   "outputs": [],
   "source": [
    "study.sources = [Source(name=\"GSM255770\"), Source(name=\"GSM255771\"),Source(name=\"GSM255772\"),Source(name=\"GSM255773\")]\n",
    "study.samples = [Sample(name=\"GSM255770\"), Sample(name=\"GSM255771\"), Sample(name=\"GSM255772\"), Sample(name=\"GSM255773\"),Sample(name=\"GSM255770.a1\"), Sample(name=\"GSM255771.a1\"), Sample(name=\"GSM255772.a1\"), Sample(name=\"GSM255773.a1\")]\n",
    "\n",
    "# Note how the treatment groups are defined as sets of factor values attached to the ISA.Sample object\n",
    "study.samples[0].factor_values=[fv1,fv2,fv4]\n",
    "study.samples[1].factor_values=[fv1,fv3,fv4]\n",
    "study.samples[2].factor_values=[fv1,fv2,fv5]\n",
    "study.samples[3].factor_values=[fv1,fv3,fv5]\n",
    "\n",
    "characteristic_organism = Characteristic(category=OntologyAnnotation(term=\"Organism\"),\n",
    "                                     value=OntologyAnnotation(term=\"marine metagenome\", term_source=ncbitaxon,\n",
    "                                                              term_accession=\"http://purl.obolibrary.org/obo/NCBITaxon_408172\"))\n",
    "\n",
    "\n",
    "# Now creating a Process showing a `Protocol Application` using Source as input and producing Sample as output.\n",
    "for i in range(len(study.sources)):\n",
    "    \n",
    "    study.sources[i].characteristics.append(characteristic_organism)\n",
    "\n",
    "\n",
    "    study.process_sequence.append(Process(executes_protocol=study.protocols[0],\n",
    "                                         inputs=[study.sources[i]],\n",
    "                                         outputs=[study.samples[i]])\n",
    "                                                  )\n",
    "\n",
    "    \n",
    "    study.process_sequence.append(Process(executes_protocol=study.protocols[2],\n",
    "                                         inputs=[study.samples[i]],\n",
    "                                         outputs=[study.samples[i+4]])\n",
    "                                                  )\n",
    "# Now appending the ISA Study object to the ISA Investigation object    \n",
    "investigation.studies = [study]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "481e927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay = Assay(filename=\"a_gilbert-assay-Gx.txt\")\n",
    "assay.measurement_type = OntologyAnnotation(term=\"metagenome sequencing\",term_accession=\"http://purl.obolibrary.org/obo/OBI_0002623\", term_source=obi)\n",
    "assay.technology_type = OntologyAnnotation(term=\"nucleotide sequencing\", term_accession=\"http://purl.obolibrary.org/obo/OBI_0000626\", term_source=obi)\n",
    "# assay.technology_type = OntologyAnnotation(term=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd9df27",
   "metadata": {},
   "outputs": [],
   "source": [
    "aliquots = [Sample(name=\"GSM255770.a1\"), Sample(name=\"GSM255771.a1\"), Sample(name=\"GSM255772.a1\"), Sample(name=\"GSM255773.a1\")] \n",
    "\n",
    "for i, sample in enumerate(aliquots):\n",
    "\n",
    "    # create an extraction process that executes the extraction protocol\n",
    "\n",
    "    extraction_process = Process(executes_protocol=study.protocols[1])\n",
    "\n",
    "    # extraction process takes as input a sample, and produces an extract material as output\n",
    "    \n",
    "    char_ext = Characteristic(category=OntologyAnnotation(term=\"Material Type\"),\n",
    "                                     value=OntologyAnnotation(term=\"pellet\"))\n",
    "\n",
    "    extraction_process.inputs.append(sample)\n",
    "    material = Material(name=\"extract-{}\".format(i))\n",
    "    material.type = \"Extract Name\"\n",
    "    material.characteristics.append(char_ext)\n",
    "    extraction_process.outputs.append(material)\n",
    "\n",
    "    # create a sequencing process that executes the sequencing protocol\n",
    "\n",
    "    sequencing_process = Process(executes_protocol=study.protocols[6])\n",
    "    sequencing_process.name = \"assay-name-{}\".format(i)\n",
    "    sequencing_process.inputs.append(extraction_process.outputs[0])\n",
    "#     sequencing_process.inputs.append(material)\n",
    "\n",
    "    # Sequencing process usually has an output data file\n",
    "\n",
    "    datafile = DataFile(filename=\"sequenced-data-{}\".format(i), label=\"Raw Data File\")\n",
    "    data_comment = Comment(name=\"data_comment\",value=\"data_value\")\n",
    "    datafile.comments.append(data_comment)\n",
    "    sequencing_process.outputs.append(datafile)\n",
    "\n",
    "    # Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set\n",
    "    # these links for you. It is found in the isatools.model package\n",
    "\n",
    "    plink(extraction_process, sequencing_process)\n",
    "\n",
    "    # make sure the extract, data file, and the processes are attached to the assay\n",
    "\n",
    "\n",
    "    assay.samples.append(sample)\n",
    "    assay.other_material.append(material)\n",
    "    assay.data_files.append(datafile)\n",
    "    \n",
    "    assay.process_sequence.append(extraction_process)\n",
    "    assay.process_sequence.append(sequencing_process)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6b3d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "study.assays.append(assay)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "735aae92",
   "metadata": {},
   "outputs": [],
   "source": [
    "from isatools.isatab import dump\n",
    "\n",
    "# note the use of the flag for explicit serialization on factor values on assay tables\n",
    "dump(investigation, \"./output/BII-S-9/\", write_factor_values_in_assay_table=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d38d02",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isa-api-py39",
   "language": "python",
   "name": "isa-api-py39"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
