
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>&lt;no title&gt; &#8212; The ISA cookbook</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet" />
  <link href="../../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.5f77b4aec8189eecf79907ce328c390d.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/mermaid.js"></script>
    <script src="../../_static/mermaid.min.js"></script>
    <script src="../../_static/mermaid.core.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "isa-tools/isa-api");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "ðŸ’¬ comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="canonical" href="https://isa-tools.github.io/isa-cookbook/content/notebooks/ISA-API -More than one Sample Name field in assay tables.html" />
    <link rel="shortcut icon" href="../../_static/isa.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      <img src="../../_static/isa_footer_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">The ISA cookbook</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search for FAIR recipes..." aria-label="Search for FAIR recipes..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   ISA tools API
  </a>
 </li>
</ul>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../installation.html">
   ISA-API installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../isamodel.html">
   ISA data model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../reading.html">
   Reading in ISA-Tab or ISA JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../creation.html">
   Creating ISA content
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../creationtutorial.html">
   Describing a simple experiment with
   <code class="docutils literal notranslate">
    <span class="pre">
     isa
    </span>
    <span class="pre">
     objects
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../validation.html">
   Validating ISA-Tab and ISA-JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conversions.html">
   Converting between ISA format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../importdata.html">
   Importing data into ISA formats
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../exportdata.html">
   Exporting data from ISA formats
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../examples.html">
   ISA API example programs
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-createSimpleISAtab.html">
     createSimpleISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-createSimpleISAJSON.html">
     createSimpleISAjson.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-validateISAtab.html">
     validateISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-validateISAjson.html">
     validateISAjson.py
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../github.html">
   Downloading files stored in Github
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../knownissues.html">
   Release Notes
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../notebooks.html">
   Hands-on Practicals
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="reading-isa-tab-from-disk-exploring-and-validating.html">
     Reading ISA-Tab from files and Validating ISA-Tab files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="programmatic-rebuild-of-BII-S-3-with-ISA-API.html">
     Programmatic recreation of BII-S-3 with the ISA-API
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="isa-protocol-type-synonym-effect-on-assay-table.html">
     ISA-API built-in semantic requirements for
     <code class="docutils literal notranslate">
      <span class="pre">
       protocol
      </span>
      <span class="pre">
       type
      </span>
     </code>
     values
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create-a-simple-ISA-descriptor.html">
     Create a simple ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create_2x2x2_factorial_design.html">
     Create an ISA Study representing a 2x2x2 factorial design with single measurements
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create-a-repeated-treatment-design-with-ISA-descriptor.html">
     Create a repeated treatment design with ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="light-sensitity-study-repeated-treatment-design.html">
     Light Sensitivity Experiment:
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="isa-metabolomics-study-design-create.html">
     ISA Create Mode example
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="investigation-from-datascriptor-config-observational-variables-and-ontology-annotation.html">
     Create ISA-API Investigation from Datascriptor Study Design configuration
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="isa-tab-unused-sources-not-serialized-brapi.html">
     Known Issue: unused ISA Source arenâ€™t serialized to ISA-Tab
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/content/notebooks/ISA-API -More than one Sample Name field in assay tables.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/isa-tools/isa-api"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/isa-tools/isa-api/issues/new?title=Issue%20on%20page%20%2Fcontent/notebooks/ISA-API -More than one Sample Name field in assay tables.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/isa-tools/isa-api/edit/master/isa-cookbook/content/notebooks/ISA-API -More than one Sample Name field in assay tables.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/isa-tools/isa-api/master?urlpath=lab/tree/isa-cookbook/content/notebooks/ISA-API -More than one Sample Name field in assay tables.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/isa-tools/isa-api/blob/master/isa-cookbook/content/notebooks/ISA-API -More than one Sample Name field in assay tables.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="simple visible nav section-nav flex-column">
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">isatools.model</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">Comment</span><span class="p">,</span>
    <span class="n">Investigation</span><span class="p">,</span>
    <span class="n">Study</span><span class="p">,</span>
    <span class="n">StudyFactor</span><span class="p">,</span>
    <span class="n">FactorValue</span><span class="p">,</span>
    <span class="n">OntologyAnnotation</span><span class="p">,</span>
    <span class="n">Characteristic</span><span class="p">,</span>
    <span class="n">OntologySource</span><span class="p">,</span>
    <span class="n">Material</span><span class="p">,</span>
    <span class="n">Sample</span><span class="p">,</span>
    <span class="n">Source</span><span class="p">,</span>
    <span class="n">Protocol</span><span class="p">,</span>
    <span class="n">ProtocolParameter</span><span class="p">,</span>
    <span class="n">ProtocolComponent</span><span class="p">,</span>
    <span class="n">ParameterValue</span><span class="p">,</span>
    <span class="n">Process</span><span class="p">,</span>
    <span class="n">Publication</span><span class="p">,</span>
    <span class="n">Person</span><span class="p">,</span>
    <span class="n">Assay</span><span class="p">,</span>
    <span class="n">DataFile</span><span class="p">,</span>
    <span class="n">plink</span>
<span class="p">)</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">investigation</span> <span class="o">=</span> <span class="n">Investigation</span><span class="p">()</span>
<span class="c1"># i_comment = Comment(name=&quot;i_comment&quot;, value=&quot;i_value&quot;)</span>
<span class="c1"># investigation.comments.append(i_comment)</span>


<span class="c1"># Declaring the Ontologies and Vocabularies used in the ISA Study</span>
<span class="c1"># dummy_onto=OntologySource(name=&quot;Dumbo&quot;,description=&quot;&quot;)</span>
<span class="n">chebi</span><span class="o">=</span><span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;CHEBI&quot;</span><span class="p">,</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;Chemical Entity of Biological Interest&quot;</span><span class="p">)</span>
<span class="n">efo</span><span class="o">=</span><span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;EFO&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Experimental Factor Ontology&quot;</span><span class="p">)</span>
<span class="n">obi</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;OBI&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ontology for Biomedical Investigations&quot;</span><span class="p">)</span>
<span class="n">pato</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;PATO&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Phenotype and Trait Ontology&quot;</span><span class="p">)</span>
<span class="n">ncbitaxon</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;NCIBTaxon&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;NCBI Taxonomy&quot;</span><span class="p">)</span>
<span class="n">investigation</span><span class="o">.</span><span class="n">ontology_source_references</span><span class="o">=</span><span class="p">[</span><span class="n">chebi</span><span class="p">,</span><span class="n">efo</span><span class="p">,</span><span class="n">obi</span><span class="p">,</span><span class="n">pato</span><span class="p">,</span><span class="n">ncbitaxon</span><span class="p">]</span>


<span class="n">study</span> <span class="o">=</span> <span class="n">Study</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;s_BII-S-10-synthesic.txt&quot;</span><span class="p">)</span>
<span class="c1"># st_comment = Comment(name=&quot;st_comment&quot;, value=&quot;st_value&quot;)</span>
<span class="c1"># study.comments.append(st_comment)</span>
<span class="n">study</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;BII-S-10-synth&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;cross-omics synthetic experiment&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;cross-omics experiment testing ISA-API support for sample aliquoting at study or assay level&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">submission_date</span> <span class="o">=</span> <span class="s2">&quot;15/08/2021&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">public_release_date</span> <span class="o">=</span> <span class="s2">&quot;15/08/2021&quot;</span>

<span class="c1"># These NCBI SRA related ISA Comments fields are required and must be present for the ISA SRAconverter is to be invoked later</span>
<span class="n">src_comment_sra1</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SRA Broker Name&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;OXFORD&quot;</span><span class="p">)</span>
<span class="n">src_comment_sra2</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SRA Center Name&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;OXFORD&quot;</span><span class="p">)</span>
<span class="n">src_comment_sra3</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SRA Center Project Name&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;OXFORD&quot;</span><span class="p">)</span>
<span class="n">src_comment_sra4</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SRA Lab Name&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;Oxford e-Research Centre&quot;</span><span class="p">)</span>
<span class="n">src_comment_sra5</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;SRA Submission Action&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;ADD&quot;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_sra1</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_sra2</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_sra3</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_sra4</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_sra5</span><span class="p">)</span>

<span class="c1"># These ISA Comments are optional and may be used to report funding information</span>
<span class="n">src_comment_st1</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Study Funding Agency&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">src_comment_st2</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Study Grant Number&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_st1</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_comment_st2</span><span class="p">)</span>
   
<span class="c1"># Declaring all the protocols used in the ISA study. Note also the declaration of Protocol Parameters when needed.</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;environmental material collection - standard procedure 1&quot;</span><span class="p">,</span>
             <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Waters samples were prefiltered through a 1.6 um GF/A glass fibre filter to reduce Eukaryotic contamination. Filtrate was then collected on a 0.2 um Sterivex (millipore) filter which was frozen in liquid nitrogen until nucelic acid extraction. CO2 bubbled through 11000 L mesocosm to simulate ocean acidification predicted conditions. Then phosphate and nitrate were added to induce a phytoplankton bloom.&quot;</span><span class="p">,</span>
             <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">),</span>
             <span class="n">parameters</span><span class="o">=</span><span class="p">[</span>
              <span class="n">ProtocolParameter</span><span class="p">(</span><span class="n">parameter_name</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;filter pore size&quot;</span><span class="p">))</span>
             <span class="p">]</span>
            <span class="p">),</span>
    <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;aliquoting-procedure&quot;</span><span class="p">,</span>
             <span class="n">description</span><span class="o">=</span><span class="s2">&quot;aliquoting&quot;</span><span class="p">,</span>
             <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">)),</span>
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;nucleic acid extraction&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Total nucleic acid extraction was done as quickly as possible using the method of Neufeld et al, 2007.&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid extraction&quot;</span><span class="p">)</span>
    <span class="p">),</span>
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;sample aliquoting - standard procedure 3&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;splitting collected samples into aliquots&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;aliquoting&quot;</span><span class="p">)</span> 
    <span class="p">),</span>
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;genomic DNA extraction - standard procedure 4&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;superscript+random hexamer primer&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid extraction&quot;</span><span class="p">)</span>
    <span class="p">),</span>
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;reverse transcription - standard procedure 5&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;reverse transcription&quot;</span><span class="p">),</span>
    <span class="p">),</span>
     <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;library construction&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;library construction&quot;</span><span class="p">),</span>
        <span class="n">parameters</span><span class="o">=</span><span class="p">[</span>
            <span class="n">ProtocolParameter</span><span class="p">(</span><span class="n">parameter_name</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;library strategy&quot;</span><span class="p">)),</span>
            <span class="n">ProtocolParameter</span><span class="p">(</span><span class="n">parameter_name</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;library layout&quot;</span><span class="p">)),</span>
            <span class="n">ProtocolParameter</span><span class="p">(</span><span class="n">parameter_name</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;library selection&quot;</span><span class="p">))</span>
        <span class="p">]</span>
    <span class="p">),</span>   
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;nucleic acid sequencing&quot;</span><span class="p">,</span> <span class="c1">#pyrosequencing - standard procedure 6&quot;,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;1. Sample Input and Fragmentation: The Genome Sequencer FLX System supports the sequencing of samples from a wide variety of starting materials including genomic DNA, PCR products, BACs, and cDNA. Samples such as genomic DNA and BACs are fractionated into small, 300- to 800-base pair fragments. For smaller samples, such as small non-coding RNA or PCR amplicons, fragmentation is not required. Instead, short PCR products amplified using Genome Sequencer fusion primers can be used for immobilization onto DNA capture beads as shown below.&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid sequencing&quot;</span><span class="p">),</span>
        <span class="n">parameters</span><span class="o">=</span><span class="p">[</span>
            <span class="n">ProtocolParameter</span><span class="p">(</span><span class="n">parameter_name</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sequencing instrument&quot;</span><span class="p">))</span>
        <span class="p">]</span>
    <span class="p">),</span>
    <span class="n">Protocol</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;sequence analysis - standard procedure 7&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;data transformation&quot;</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">]</span>


<span class="c1"># Adding a Study Design descriptor to the ISA Study object</span>
<span class="n">intervention_design</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term_source</span><span class="o">=</span><span class="n">obi</span><span class="p">)</span>
<span class="n">intervention_design</span><span class="o">.</span><span class="n">term</span> <span class="o">=</span> <span class="s2">&quot;intervention design&quot;</span>
<span class="n">intervention_design</span><span class="o">.</span><span class="n">term_accession</span> <span class="o">=</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/OBI_0000115&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">design_descriptors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intervention_design</span><span class="p">)</span>


<span class="c1"># Declaring the Study Factors</span>
<span class="n">study</span><span class="o">.</span><span class="n">factors</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">StudyFactor</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;compound&quot;</span><span class="p">,</span><span class="n">factor_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;chemical substance&quot;</span><span class="p">,</span>
                                                              <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.obolibrary.org/obo/CHEBI_59999&quot;</span><span class="p">,</span>
                                                              <span class="n">term_source</span><span class="o">=</span><span class="n">chebi</span><span class="p">)),</span>
    <span class="n">StudyFactor</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;dose&quot;</span><span class="p">,</span><span class="n">factor_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;dose&quot;</span><span class="p">,</span> <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://www.ebi.ac.uk/efo/EFO_0000428&quot;</span><span class="p">,</span><span class="n">term_source</span><span class="o">=</span><span class="n">efo</span><span class="p">)),</span>
    <span class="n">StudyFactor</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;collection time&quot;</span><span class="p">,</span><span class="n">factor_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;time&quot;</span><span class="p">,</span> <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.obolibrary.org/obo/PATO_0000165&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">pato</span><span class="p">))</span>
<span class="p">]</span>

<span class="c1"># Associating the levels to each of the Study Factor.</span>
<span class="n">fv1</span> <span class="o">=</span> <span class="n">FactorValue</span><span class="p">(</span><span class="n">factor_name</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;atorvastatin&quot;</span><span class="p">))</span>
<span class="n">fv2</span> <span class="o">=</span> <span class="n">FactorValue</span><span class="p">(</span><span class="n">factor_name</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;high dose&quot;</span><span class="p">))</span>
<span class="n">fv3</span> <span class="o">=</span> <span class="n">FactorValue</span><span class="p">(</span><span class="n">factor_name</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;low dose&quot;</span><span class="p">))</span>
<span class="n">fv4</span> <span class="o">=</span> <span class="n">FactorValue</span><span class="p">(</span><span class="n">factor_name</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;2 months&quot;</span><span class="p">)</span>
<span class="n">fv5</span> <span class="o">=</span> <span class="n">FactorValue</span><span class="p">(</span><span class="n">factor_name</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">factors</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;18 months&quot;</span><span class="p">)</span>


<span class="c1"># Adding the publications associated to the study</span>
<span class="n">study</span><span class="o">.</span><span class="n">publications</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">Publication</span><span class="p">(</span><span class="n">doi</span><span class="o">=</span><span class="s2">&quot;10.1371/journal.pone.0003042&quot;</span><span class="p">,</span><span class="n">pubmed_id</span><span class="o">=</span><span class="s2">&quot;18725995&quot;</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Detection of large numbers of novel sequences in the metatranscriptomes of complex marine microbial communities.&quot;</span><span class="p">,</span>
                <span class="n">status</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;indexed in PubMed&quot;</span><span class="p">),</span>
                <span class="n">author_list</span><span class="o">=</span><span class="s2">&quot;Gilbert JA, Field D, Huang Y, Edwards R, Li W, Gilna P, Joint I.&quot;</span><span class="p">)</span>  
<span class="p">]</span>


<span class="c1"># Adding the authors of the study</span>
<span class="n">study</span><span class="o">.</span><span class="n">contacts</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">Person</span><span class="p">(</span><span class="n">first_name</span><span class="o">=</span><span class="s2">&quot;Rex&quot;</span><span class="p">,</span> <span class="n">last_name</span><span class="o">=</span><span class="s2">&quot;Durand&quot;</span><span class="p">,</span> <span class="n">affiliation</span><span class="o">=</span><span class="s2">&quot;LHC Laboratory&quot;</span><span class="p">,</span> <span class="n">email</span><span class="o">=</span><span class="s2">&quot;rex.durand@lhcl.ac.uk&quot;</span><span class="p">,</span>
           <span class="n">address</span><span class="o">=</span><span class="s2">&quot;Nevsky perspective, Bournemouth, United Kingdom&quot;</span><span class="p">,</span>
           <span class="n">comments</span><span class="o">=</span><span class="p">[</span><span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Study Person REF&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)],</span>
            <span class="n">roles</span><span class="o">=</span><span class="p">[</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;principal investigator role&quot;</span><span class="p">),</span>
                   <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;SRA Inform On Status&quot;</span><span class="p">),</span>
                   <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;SRA Inform On Error&quot;</span><span class="p">)]</span>
    <span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">study</span><span class="o">.</span><span class="n">sources</span> <span class="o">=</span> <span class="p">[</span><span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255770&quot;</span><span class="p">),</span> <span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255771&quot;</span><span class="p">),</span><span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255772&quot;</span><span class="p">),</span><span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255773&quot;</span><span class="p">)]</span>
<span class="n">study</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255770&quot;</span><span class="p">),</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255771&quot;</span><span class="p">),</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255772&quot;</span><span class="p">),</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;GSM255773&quot;</span><span class="p">)]</span>

<span class="c1"># Note how the treatment groups are defined as sets of factor values attached to the ISA.Sample object</span>
<span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">factor_values</span><span class="o">=</span><span class="p">[</span><span class="n">fv1</span><span class="p">,</span><span class="n">fv2</span><span class="p">,</span><span class="n">fv4</span><span class="p">]</span>
<span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">factor_values</span><span class="o">=</span><span class="p">[</span><span class="n">fv1</span><span class="p">,</span><span class="n">fv3</span><span class="p">,</span><span class="n">fv4</span><span class="p">]</span>
<span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">factor_values</span><span class="o">=</span><span class="p">[</span><span class="n">fv1</span><span class="p">,</span><span class="n">fv2</span><span class="p">,</span><span class="n">fv5</span><span class="p">]</span>
<span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">factor_values</span><span class="o">=</span><span class="p">[</span><span class="n">fv1</span><span class="p">,</span><span class="n">fv3</span><span class="p">,</span><span class="n">fv5</span><span class="p">]</span>

<span class="n">characteristic_organism</span> <span class="o">=</span> <span class="n">Characteristic</span><span class="p">(</span><span class="n">category</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Organism&quot;</span><span class="p">),</span>
                                     <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;marine metagenome&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">ncbitaxon</span><span class="p">,</span>
                                                              <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.obolibrary.org/obo/NCBITaxon_408172&quot;</span><span class="p">))</span>


<span class="c1"># Now creating a Process showing a `Protocol Application` using Source as input and producing Sample as output.</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="p">)):</span>
    
    <span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">characteristics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">characteristic_organism</span><span class="p">)</span>


    <span class="n">study</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                         <span class="n">inputs</span><span class="o">=</span><span class="p">[</span><span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span>
                                         <span class="n">outputs</span><span class="o">=</span><span class="p">[</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
                                                  <span class="p">)</span>


<span class="c1"># Now appending the ISA Study object to the ISA Investigation object    </span>
<span class="n">investigation</span><span class="o">.</span><span class="n">studies</span> <span class="o">=</span> <span class="p">[</span><span class="n">study</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">assay</span> <span class="o">=</span> <span class="n">Assay</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;a_gilbert-assay-Gx.txt&quot;</span><span class="p">)</span>
<span class="n">assay</span><span class="o">.</span><span class="n">measurement_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;metagenome sequencing&quot;</span><span class="p">,</span><span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.obolibrary.org/obo/OBI_0002623&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">obi</span><span class="p">)</span>
<span class="n">assay</span><span class="o">.</span><span class="n">technology_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleotide sequencing&quot;</span><span class="p">,</span> <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.obolibrary.org/obo/OBI_0000626&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">obi</span><span class="p">)</span>
<span class="c1"># assay.technology_type = OntologyAnnotation(term=&quot;&quot;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># aliquots = [Sample(name=&quot;GSM255770.a1&quot;), Sample(name=&quot;GSM255771.a1&quot;), Sample(name=&quot;GSM255772.a1&quot;), Sample(name=&quot;GSM255773.a1&quot;)] </span>

<span class="c1"># study.samples.append(Sample(name=&quot;GSM255770.a1&quot;))</span>
<span class="c1"># study.samples.append(Sample(name=&quot;GSM255772.a1&quot;))</span>
<span class="c1"># study.samples.append(Sample(name=&quot;GSM255773.a1&quot;))</span>
<span class="c1"># study.samples.append(Sample(name=&quot;GSM255774.a1&quot;))</span>
                     
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
      
    <span class="c1"># create an aliquoting process which creates new samples from existing samples created in a study.processSequence</span>
    <span class="n">aliquoting_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">aliquoting_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">aliquot</span> <span class="o">=</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;aliquot-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span><span class="n">derives_from</span><span class="o">=</span><span class="p">[</span><span class="n">sample</span><span class="p">])</span>
   
    <span class="n">char_alq</span> <span class="o">=</span> <span class="n">Characteristic</span><span class="p">(</span><span class="n">category</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Material&quot;</span><span class="p">),</span>
                                     <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;aliquot&quot;</span><span class="p">))</span>
    <span class="n">aliquot</span><span class="o">.</span><span class="n">characteristics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">char_alq</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">aliquot</span><span class="p">)</span>
    

    <span class="n">aliquoting_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aliquot</span><span class="p">)</span>                       
    
    <span class="n">assay</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aliquot</span><span class="p">)</span>
    
    <span class="c1"># create an extraction process that executes the extraction protocol </span>
    <span class="n">extraction_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="c1"># extraction process takes as input a sample, and produces an extract material as output   </span>
    <span class="n">char_ext</span> <span class="o">=</span> <span class="n">Characteristic</span><span class="p">(</span><span class="n">category</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Material Type&quot;</span><span class="p">),</span>
                                     <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;pellet&quot;</span><span class="p">))</span>

    <span class="n">extraction_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aliquot</span><span class="p">)</span>
    <span class="n">material</span> <span class="o">=</span> <span class="n">Material</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;extract-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">material</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Extract Name&quot;</span>
    <span class="n">material</span><span class="o">.</span><span class="n">characteristics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">char_ext</span><span class="p">)</span>
    <span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>

    <span class="c1"># create a sequencing process that executes the sequencing protocol</span>

    <span class="n">sequencing_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;assay-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1"># sequencing_process.inputs.append(material)</span>

    <span class="c1"># Sequencing process usually has an output data file</span>

    <span class="n">datafile</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sequenced-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Raw Data File&quot;</span><span class="p">)</span>
    <span class="n">data_comment</span> <span class="o">=</span> <span class="n">Comment</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_comment&quot;</span><span class="p">,</span><span class="n">value</span><span class="o">=</span><span class="s2">&quot;data_value&quot;</span><span class="p">)</span>
    <span class="n">datafile</span><span class="o">.</span><span class="n">comments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data_comment</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>



    <span class="c1"># make sure the extract, data file, and the processes are attached to the assay</span>


    <span class="n">assay</span><span class="o">.</span><span class="n">other_material</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">data_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    
    <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aliquoting_process</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">)</span>

    <span class="c1"># Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set</span>
    <span class="c1"># these links for you. It is found in the isatools.model package</span>

    <span class="n">plink</span><span class="p">(</span><span class="n">aliquoting_process</span><span class="p">,</span> <span class="n">sequencing_process</span><span class="p">)</span>
<span class="c1">#     plink(aliquoting_process, extraction_process)</span>
<span class="c1">#     plink(extraction_process, sequencing_process)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample(
    name=aliquot-0
    characteristics=1 Characteristic objects
    factor_values=0 FactorValue objects
    derives_from=1 Source objects
    comments=0 Comment objects
)
Sample(
    name=aliquot-1
    characteristics=1 Characteristic objects
    factor_values=0 FactorValue objects
    derives_from=1 Source objects
    comments=0 Comment objects
)
Sample(
    name=aliquot-2
    characteristics=1 Characteristic objects
    factor_values=0 FactorValue objects
    derives_from=1 Source objects
    comments=0 Comment objects
)
Sample(
    name=aliquot-3
    characteristics=1 Characteristic objects
    factor_values=0 FactorValue objects
    derives_from=1 Source objects
    comments=0 Comment objects
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span>

<span class="c1"># for s in range(len(study.assays[0].samples)):</span>
<span class="c1">#     print(study.assays[0].samples[s])</span>
    
<span class="nb">print</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample(
    name=aliquot-3
    characteristics=1 Characteristic objects
    factor_values=0 FactorValue objects
    derives_from=1 Source objects
    comments=0 Comment objects
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">isatools.isatab</span> <span class="kn">import</span> <span class="n">dump</span>

<span class="c1"># note the use of the flag for explicit serialization on factor values on assay tables</span>
<span class="n">dump</span><span class="p">(</span><span class="n">investigation</span><span class="p">,</span> <span class="s2">&quot;./output/BII-S-10/&quot;</span><span class="p">,</span> <span class="n">write_factor_values_in_assay_table</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2021-08-02 18:39:47,234 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [616, 617, 618, 619]
2021-08-02 18:39:47,237 [WARNING]: isatab.py(write_study_table_files:1194) &gt;&gt; [624, 620, 616, 625, 621, 617, 626, 622, 618, 627, 623, 619]
2021-08-02 18:39:47,238 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[616, 624, 620], [617, 625, 621], [618, 626, 622], [619, 627, 623]]
2021-08-02 18:39:47,281 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [629, 620, 635, 621, 641, 622, 647, 623]
2021-08-02 18:39:47,282 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[629, 630, 631, 632], [629, 630], [620, 628, 629, 630], [620, 628, 629, 630, 631, 632], [635, 636], [635, 636, 637, 638], [621, 634, 635, 636], [621, 634, 635, 636, 637, 638], [641, 642], [641, 642, 643, 644], [622, 640, 641, 642], [622, 640, 641, 642, 643, 644], [647, 648], [647, 648, 649, 650], [623, 646, 647, 648], [623, 646, 647, 648, 649, 650]]
2021-08-02 18:39:47,282 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[629, 630, 631, 632], [629, 630], [620, 628, 629, 630], [620, 628, 629, 630, 631, 632], [635, 636], [635, 636, 637, 638], [621, 634, 635, 636], [621, 634, 635, 636, 637, 638], [641, 642], [641, 642, 643, 644], [622, 640, 641, 642], [622, 640, 641, 642, 643, 644], [647, 648], [647, 648, 649, 650], [623, 646, 647, 648], [623, 646, 647, 648, 649, 650]]
2021-08-02 18:39:47,297 [CRITICAL]: isatab.py(write_assay_table_files:1554) &gt;&gt; Error thrown: column labels are: Index([&#39;Sample Name&#39;, &#39;Protocol REF.aliquoting-procedure&#39;, &#39;Sample Name&#39;,
       &#39;Protocol REF.aliquoting-procedure&#39;, &#39;Extract Name&#39;,
       &#39;Extract Name.Characteristics[Material Type]&#39;,
       &#39;Extract Name.Characteristics[Material Type].Term Source REF&#39;,
       &#39;Extract Name.Characteristics[Material Type].Term Accession Number&#39;,
       &#39;Protocol REF.nucleic acid sequencing&#39;, &#39;Assay Name&#39;, &#39;Raw Data File&#39;,
       &#39;Raw Data File.Comment[data_comment]&#39;],
      dtype=&#39;object&#39;)
2021-08-02 18:39:47,306 [CRITICAL]: isatab.py(write_assay_table_files:1555) &gt;&gt; Error thrown: data is:    Sample Name Protocol REF.aliquoting-procedure Sample Name  \
0    aliquot-0              aliquoting-procedure   aliquot-0   
1    aliquot-0              aliquoting-procedure   aliquot-0   
2    aliquot-0              aliquoting-procedure   aliquot-0   
3    aliquot-0              aliquoting-procedure   aliquot-0   
4    aliquot-1              aliquoting-procedure   aliquot-1   
5    aliquot-1              aliquoting-procedure   aliquot-1   
6    aliquot-1              aliquoting-procedure   aliquot-1   
7    aliquot-1              aliquoting-procedure   aliquot-1   
8    aliquot-2              aliquoting-procedure   aliquot-2   
9    aliquot-2              aliquoting-procedure   aliquot-2   
10   aliquot-2              aliquoting-procedure   aliquot-2   
11   aliquot-2              aliquoting-procedure   aliquot-2   
12   aliquot-3              aliquoting-procedure   aliquot-3   
13   aliquot-3              aliquoting-procedure   aliquot-3   
14   aliquot-3              aliquoting-procedure   aliquot-3   
15   aliquot-3              aliquoting-procedure   aliquot-3   

   Protocol REF.aliquoting-procedure Extract Name  \
0               aliquoting-procedure    extract-0   
1               aliquoting-procedure                
2               aliquoting-procedure                
3               aliquoting-procedure    extract-0   
4               aliquoting-procedure                
5               aliquoting-procedure    extract-1   
6               aliquoting-procedure                
7               aliquoting-procedure    extract-1   
8               aliquoting-procedure                
9               aliquoting-procedure    extract-2   
10              aliquoting-procedure                
11              aliquoting-procedure    extract-2   
12              aliquoting-procedure                
13              aliquoting-procedure    extract-3   
14              aliquoting-procedure                
15              aliquoting-procedure    extract-3   

   Extract Name.Characteristics[Material Type]  \
0                                       pellet   
1                                                
2                                                
3                                       pellet   
4                                                
5                                       pellet   
6                                                
7                                       pellet   
8                                                
9                                       pellet   
10                                               
11                                      pellet   
12                                               
13                                      pellet   
14                                               
15                                      pellet   

   Extract Name.Characteristics[Material Type].Term Source REF  \
0                                                                
1                                                                
2                                                                
3                                                                
4                                                                
5                                                                
6                                                                
7                                                                
8                                                                
9                                                                
10                                                               
11                                                               
12                                                               
13                                                               
14                                                               
15                                                               

   Extract Name.Characteristics[Material Type].Term Accession Number  \
0                                                                      
1                                                                      
2                                                                      
3                                                                      
4                                                                      
5                                                                      
6                                                                      
7                                                                      
8                                                                      
9                                                                      
10                                                                     
11                                                                     
12                                                                     
13                                                                     
14                                                                     
15                                                                     

   Protocol REF.nucleic acid sequencing    Assay Name     Raw Data File  \
0               nucleic acid sequencing  assay-name-0  sequenced-data-0   
1                                                                         
2                                                                         
3               nucleic acid sequencing  assay-name-0  sequenced-data-0   
4                                                                         
5               nucleic acid sequencing  assay-name-1  sequenced-data-1   
6                                                                         
7               nucleic acid sequencing  assay-name-1  sequenced-data-1   
8                                                                         
9               nucleic acid sequencing  assay-name-2  sequenced-data-2   
10                                                                        
11              nucleic acid sequencing  assay-name-2  sequenced-data-2   
12                                                                        
13              nucleic acid sequencing  assay-name-3  sequenced-data-3   
14                                                                        
15              nucleic acid sequencing  assay-name-3  sequenced-data-3   

   Raw Data File.Comment[data_comment]  
0                           data_value  
1                                       
2                                       
3                           data_value  
4                                       
5                           data_value  
6                                       
7                           data_value  
8                                       
9                           data_value  
10                                      
11                          data_value  
12                                      
13                          data_value  
14                                      
15                          data_value  
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">5</span><span class="n">n</span><span class="o">/</span><span class="n">rl6lqnks4rqb59pbtpvvntqw0000gr</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">ipykernel_71489</span><span class="o">/</span><span class="mf">407357070.</span><span class="n">py</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> 
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="c1"># note the use of the flag for explicit serialization on factor values on assay tables</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">dump</span><span class="p">(</span><span class="n">investigation</span><span class="p">,</span> <span class="s2">&quot;./output/BII-S-10/&quot;</span><span class="p">,</span> <span class="n">write_factor_values_in_assay_table</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nn">~/.pyenv/versions/3.9.0/envs/isa-api-py39/src/isatools/isatools/isatab.py</span> in <span class="ni">dump</span><span class="nt">(isa_obj, output_path, i_file_name, skip_dump_tables, write_factor_values_in_assay_table)</span>
<span class="g g-Whitespace">   </span><span class="mi">1048</span>     <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1049</span>         <span class="n">write_study_table_files</span><span class="p">(</span><span class="n">investigation</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1050</span>         <span class="n">write_assay_table_files</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1051</span>             <span class="n">investigation</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">write_factor_values_in_assay_table</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1052</span> 

<span class="nn">~/.pyenv/versions/3.9.0/envs/isa-api-py39/src/isatools/isatools/isatab.py</span> in <span class="ni">write_assay_table_files</span><span class="nt">(inv_obj, output_dir, write_factor_values)</span>
<span class="g g-Whitespace">   </span><span class="mi">1554</span>                 <span class="n">log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;Error thrown: column labels are: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">DF</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
<span class="g g-Whitespace">   </span><span class="mi">1555</span>                 <span class="n">log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;Error thrown: data is: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">DF</span><span class="p">))</span>
<span class="ne">-&gt; </span><span class="mi">1556</span>                 <span class="k">raise</span> <span class="n">e</span>
<span class="g g-Whitespace">   </span><span class="mi">1557</span>             <span class="c1"># arbitrary sort on column 0</span>
<span class="g g-Whitespace">   </span><span class="mi">1558</span> 

<span class="nn">~/.pyenv/versions/3.9.0/envs/isa-api-py39/src/isatools/isatools/isatab.py</span> in <span class="ni">write_assay_table_files</span><span class="nt">(inv_obj, output_dir, write_factor_values)</span>
<span class="g g-Whitespace">   </span><span class="mi">1550</span>             <span class="n">DF</span> <span class="o">=</span> <span class="n">DF</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span>  <span class="c1"># reorder columns</span>
<span class="g g-Whitespace">   </span><span class="mi">1551</span>             <span class="k">try</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1552</span>                 <span class="n">DF</span> <span class="o">=</span> <span class="n">DF</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">DF</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1553</span>             <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1554</span>                 <span class="n">log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;Error thrown: column labels are: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">DF</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>

<span class="nn">~/.pyenv/versions/3.9.0/envs/isa-api-py39/lib/python3.9/site-packages/pandas/core/frame.py</span> in <span class="ni">sort_values</span><span class="nt">(self, by, axis, ascending, inplace, kind, na_position, ignore_index, key)</span>
<span class="g g-Whitespace">   </span><span class="mi">5453</span> 
<span class="g g-Whitespace">   </span><span class="mi">5454</span>             <span class="n">by</span> <span class="o">=</span> <span class="n">by</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="ne">-&gt; </span><span class="mi">5455</span>             <span class="n">k</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_label_or_level_values</span><span class="p">(</span><span class="n">by</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">5456</span> 
<span class="g g-Whitespace">   </span><span class="mi">5457</span>             <span class="c1"># need to rewrap column in Series to apply key function</span>

<span class="nn">~/.pyenv/versions/3.9.0/envs/isa-api-py39/lib/python3.9/site-packages/pandas/core/generic.py</span> in <span class="ni">_get_label_or_level_values</span><span class="nt">(self, key, axis)</span>
<span class="g g-Whitespace">   </span><span class="mi">1697</span> 
<span class="g g-Whitespace">   </span><span class="mi">1698</span>             <span class="n">label_axis_name</span> <span class="o">=</span> <span class="s2">&quot;column&quot;</span> <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;index&quot;</span>
<span class="ne">-&gt; </span><span class="mi">1699</span>             <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1700</span>                 <span class="sa">f</span><span class="s2">&quot;The </span><span class="si">{</span><span class="n">label_axis_name</span><span class="si">}</span><span class="s2"> label &#39;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">&#39; is not unique.</span><span class="si">{</span><span class="n">multi_message</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1701</span>             <span class="p">)</span>

<span class="ne">ValueError</span>: The column label &#39;Sample Name&#39; is not unique.
</pre></div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "isa-api-py39"
        },
        kernelOptions: {
            kernelName: "isa-api-py39",
            path: "./content/notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'isa-api-py39'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The ISA Team<br/>
        
            &copy; Copyright 2021 (CC-BY 4.0).<br/>
          <div class="extra_footer">
            <script>mermaid.init();</script> <div class="row_footer"> This page was created thanks to work by <a href="https://github.com/jupyter/jupyter-book/graphs/contributors">The Jupyter Book Community</a></div> <div class="row_footer"> This project is supported financially by the <a href="https://wellcome.org/">Welcome Trust</a>, Elixir-UK and the European Union</div>
<div class="row_footer"> <div class="column" style="center"><a href="https://wellcome.org/"><img src="../_static/images/logo/wellcome-logo-black.png" height="25" alt="Wellcome" style="float:center""></a></div> <div class="column"><a href="https://www.elixir-uk.org.eu/"><img src="../_static/images/logo/elixir-uk_transparent.png" height="40" alt="IMI logo" style="float:center""></a></div> <div class="column"><a href="https://https://ec.europa.eu/programmes/horizon2020/en"><img src="../_static/images/logo/eu-flag-recolored.png" height="25" alt="EFPIA logo" style="float:center""></a></div> </div>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>