
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ISA-API built-in semantic requirements for protocol type values &#8212; The ISA cookbook</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet" />
  <link href="../../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.5f77b4aec8189eecf79907ce328c390d.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/mermaid.js"></script>
    <script src="../../_static/mermaid.min.js"></script>
    <script src="../../_static/mermaid.core.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "isa-tools/isa-api");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "ðŸ’¬ comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="canonical" href="https://isa-tools.github.io/isa-cookbook/content/notebooks/isa-protocol-type-synonym-effect-on-assay-table.html" />
    <link rel="shortcut icon" href="../../_static/isa.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Create a simple ISA descriptor" href="create-a-simple-ISA-descriptor.html" />
    <link rel="prev" title="Programmatic recreation of BII-S-3 with the ISA-API" href="programmatic-rebuild-of-BII-S-3-with-ISA-API.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      <img src="../../_static/isa_footer_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">The ISA cookbook</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search for FAIR recipes..." aria-label="Search for FAIR recipes..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   ISA tools API
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../installation.html">
   ISA-API installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../isamodel.html">
   ISA data model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../reading.html">
   Reading in ISA-Tab or ISA JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../creation.html">
   Creating ISA content
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../creationtutorial.html">
   Describing a simple experiment with
   <code class="docutils literal notranslate">
    <span class="pre">
     isa
    </span>
    <span class="pre">
     objects
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../validation.html">
   Validating ISA-Tab and ISA-JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conversions.html">
   Converting between ISA format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../importdata.html">
   Importing data into ISA formats
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../exportdata.html">
   Exporting data from ISA formats
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../examples.html">
   ISA API example programs
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-createSimpleISAtab.html">
     createSimpleISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-createSimpleISAJSON.html">
     createSimpleISAjson.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-validateISAtab.html">
     validateISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../examples/example-validateISAjson.html">
     validateISAjson.py
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../github.html">
   Downloading files stored in Github
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../knownissues.html">
   Release Notes
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../notebooks.html">
   Hands-on Practicals
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="reading-isa-tab-from-disk-exploring-and-validating.html">
     Reading ISA-Tab from files and Validating ISA-Tab files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="programmatic-rebuild-of-BII-S-3-with-ISA-API.html">
     Programmatic recreation of BII-S-3 with the ISA-API
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     ISA-API built-in semantic requirements for
     <code class="docutils literal notranslate">
      <span class="pre">
       protocol
      </span>
      <span class="pre">
       type
      </span>
     </code>
     values
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create-a-simple-ISA-descriptor.html">
     Create a simple ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create_2x2x2_factorial_design.html">
     Create an ISA Study representing a 2x2x2 factorial design with single measurements
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="create-a-repeated-treatment-design-with-ISA-descriptor.html">
     Create a repeated treatment design with ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="light-sensitity-study-repeated-treatment-design.html">
     Light Sensitivity Experiment:
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="isa-metabolomics-study-design-create.html">
     ISA Create Mode example
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="investigation-from-datascriptor-config-observational-variables-and-ontology-annotation.html">
     Create ISA-API Investigation from Datascriptor Study Design configuration
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="isa-tab-unused-sources-not-serialized-brapi.html">
     Known Issue: unused ISA Source arenâ€™t serialized to ISA-Tab
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/content/notebooks/isa-protocol-type-synonym-effect-on-assay-table.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/isa-tools/isa-api"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/isa-tools/isa-api/issues/new?title=Issue%20on%20page%20%2Fcontent/notebooks/isa-protocol-type-synonym-effect-on-assay-table.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/isa-tools/isa-api/edit/master/isa-cookbook/content/notebooks/isa-protocol-type-synonym-effect-on-assay-table.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/isa-tools/isa-api/master?urlpath=lab/tree/isa-cookbook/content/notebooks/isa-protocol-type-synonym-effect-on-assay-table.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/isa-tools/isa-api/blob/master/isa-cookbook/content/notebooks/isa-protocol-type-synonym-effect-on-assay-table.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#abstract">
   Abstract:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#let-s-get-the-tools">
   Letâ€™s get the tools
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-an-empty-investigation-object-and-set-some-values-to-the-instance-variables">
   1. Create an empty Investigation object and set some values to the instance variables.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-the-isa-study-graph">
   2. Create the ISA Study graph
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#building-an-assay-object-and-attach-two-protocols-extraction-and-sequencing">
   3. Building an Assay object and attach two protocols, extraction and sequencing.
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#inspecting-the-different-assay-tables-reveals-the-importance-of-using-the-right-strings">
   Inspecting the different assay tables reveals the importance of using the right strings
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#about-this-notebook">
   About this notebook
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="isa-api-built-in-semantic-requirements-for-protocol-type-values">
<h1>ISA-API built-in semantic requirements for <code class="docutils literal notranslate"><span class="pre">protocol</span> <span class="pre">type</span></code> values<a class="headerlink" href="#isa-api-built-in-semantic-requirements-for-protocol-type-values" title="Permalink to this headline">Â¶</a></h1>
<div class="section" id="abstract">
<h2>Abstract:<a class="headerlink" href="#abstract" title="Permalink to this headline">Â¶</a></h2>
<p>In this notebook, we document the consequences of not using the right <code class="docutils literal notranslate"><span class="pre">protocol</span> <span class="pre">type</span></code> value in the context of specific ISA assays.</p>
<p>It is therefore important that developers and users of the ISA-API be aware of the defautl ISA configuration and the list of protocol types used in the assay workflow associated with a given ISA Assay type, which is defined by a combination of <code class="docutils literal notranslate"><span class="pre">Measurement</span> <span class="pre">Type</span></code> and <code class="docutils literal notranslate"><span class="pre">Technology</span> <span class="pre">Type</span></code>.</p>
<p>We have documented the full list of protocol types in the following document <a class="reference external" href="https://www.todo.org">link to document</a></p>
</div>
<div class="section" id="let-s-get-the-tools">
<h2>Letâ€™s get the tools<a class="headerlink" href="#let-s-get-the-tools" title="Permalink to this headline">Â¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># If executing the notebooks on `Google Colab`,uncomment the following command </span>
<span class="c1"># and run it to install the required python libraries. Also, make the test datasets available.</span>

<span class="c1"># !pip install -r requirements.txt</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">isatools</span>
<span class="kn">from</span> <span class="nn">isatools</span> <span class="kn">import</span> <span class="n">isatab</span>

<span class="kn">from</span> <span class="nn">isatools.model</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">Investigation</span><span class="p">,</span>
    <span class="n">Study</span><span class="p">,</span>
    <span class="n">OntologySource</span><span class="p">,</span>
    <span class="n">OntologyAnnotation</span><span class="p">,</span>
    <span class="n">Person</span><span class="p">,</span>
    <span class="n">Publication</span><span class="p">,</span>
    <span class="n">Source</span><span class="p">,</span>
    <span class="n">Sample</span><span class="p">,</span>
    <span class="n">Characteristic</span><span class="p">,</span>
    <span class="n">Protocol</span><span class="p">,</span>
    <span class="n">Process</span><span class="p">,</span>
    <span class="n">Assay</span><span class="p">,</span>
    <span class="n">Material</span><span class="p">,</span>
    <span class="n">DataFile</span><span class="p">,</span>
    <span class="n">batch_create_materials</span><span class="p">,</span>
    <span class="n">plink</span>
<span class="p">)</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">isatools.isajson</span> <span class="kn">import</span> <span class="n">ISAJSONEncoder</span>
<span class="kn">from</span> <span class="nn">isatools.isatab</span> <span class="kn">import</span> <span class="n">dump_tables_to_dataframes</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-an-empty-investigation-object-and-set-some-values-to-the-instance-variables">
<h2>1. Create an empty Investigation object and set some values to the instance variables.<a class="headerlink" href="#create-an-empty-investigation-object-and-set-some-values-to-the-instance-variables" title="Permalink to this headline">Â¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">investigation</span> <span class="o">=</span> <span class="n">Investigation</span><span class="p">()</span>
<span class="c1"># investigation.identifier = &quot;i1&quot;</span>
<span class="c1"># investigation.title = &quot;My Simple ISA Investigation&quot;</span>
<span class="c1"># investigation.description = &quot;We could alternatively use the class constructor&#39;s parameters to set some default &quot; \</span>
<span class="c1">#                             &quot;values at the time of creation, however we want to demonstrate how to use the &quot; \</span>
<span class="c1">#                             &quot;object&#39;s instance variables to set values.&quot;</span>
<span class="c1"># investigation.submission_date = &quot;2016-11-03&quot;</span>
<span class="c1"># investigation.public_release_date = &quot;201611-03&quot;</span>

<span class="c1"># Create an empty Study object and set some values. The Study must have a filename, otherwise when we serialize it</span>
<span class="c1"># to ISA-Tab we would not know where to write it. We must also attach the study to the investigation by adding it</span>
<span class="c1"># to the &#39;investigation&#39; object&#39;s list of studies.</span>

<span class="n">study</span> <span class="o">=</span> <span class="n">Study</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;s_study.txt&quot;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;s1&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;My ISA Study&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;Like with the Investigation, we could use the class constructor to set some default values, &quot;</span> \
                    <span class="s2">&quot;but have chosen to demonstrate in this example the use of instance variables to set initial &quot;</span> \
                    <span class="s2">&quot;values.&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">submission_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">public_release_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>
<span class="n">investigation</span><span class="o">.</span><span class="n">studies</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">study</span><span class="p">)</span>

<span class="c1"># Some instance variables are typed with different objects and lists of objects. For example, a Study can have a</span>
<span class="c1"># list of design descriptors. A design descriptor is an Ontology Annotation describing the kind of study at hand.</span>
<span class="c1"># Ontology Annotations should typically reference an Ontology Source. We demonstrate a mix of using the class</span>
<span class="c1"># constructors and setting values with instance variables. Note that the OntologyAnnotation object</span>
<span class="c1"># &#39;intervention_design&#39; links its &#39;term_source&#39; directly to the &#39;obi&#39; object instance. To ensure the OntologySource</span>
<span class="c1"># is encapsulated in the descriptor, it is added to a list of &#39;ontology_source_references&#39; in the Investigation</span>
<span class="c1"># object. The &#39;intervention_design&#39; object is then added to the list of &#39;design_descriptors&#39; held by the Study</span>
<span class="c1"># object.</span>

<span class="n">obi</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;OBI&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ontology for Biomedical Investigations&quot;</span><span class="p">)</span>
<span class="n">investigation</span><span class="o">.</span><span class="n">ontology_source_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">obi</span><span class="p">)</span>
<span class="n">intervention_design</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term_source</span><span class="o">=</span><span class="n">obi</span><span class="p">)</span>
<span class="n">intervention_design</span><span class="o">.</span><span class="n">term</span> <span class="o">=</span> <span class="s2">&quot;intervention design&quot;</span>
<span class="n">intervention_design</span><span class="o">.</span><span class="n">term_accession</span> <span class="o">=</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/OBI_0000115&quot;</span>
<span class="n">study</span><span class="o">.</span><span class="n">design_descriptors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intervention_design</span><span class="p">)</span>

<span class="c1"># Other instance variables common to both Investigation and Study objects include &#39;contacts&#39; and &#39;publications&#39;,</span>
<span class="c1"># each with lists of corresponding Person and Publication objects.</span>

<span class="n">contact</span> <span class="o">=</span> <span class="n">Person</span><span class="p">(</span><span class="n">first_name</span><span class="o">=</span><span class="s2">&quot;Alice&quot;</span><span class="p">,</span> <span class="n">last_name</span><span class="o">=</span><span class="s2">&quot;Robertson&quot;</span><span class="p">,</span> <span class="n">affiliation</span><span class="o">=</span><span class="s2">&quot;University of Life&quot;</span><span class="p">,</span> <span class="n">roles</span><span class="o">=</span><span class="p">[</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s1">&#39;submitter&#39;</span><span class="p">)])</span>
<span class="n">study</span><span class="o">.</span><span class="n">contacts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">contact</span><span class="p">)</span>
<span class="n">publication</span> <span class="o">=</span> <span class="n">Publication</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Experiments with Elephants&quot;</span><span class="p">,</span> <span class="n">author_list</span><span class="o">=</span><span class="s2">&quot;A. Robertson, B. Robertson&quot;</span><span class="p">)</span>
<span class="n">publication</span><span class="o">.</span><span class="n">pubmed_id</span> <span class="o">=</span> <span class="s2">&quot;12345678&quot;</span>
<span class="n">publication</span><span class="o">.</span><span class="n">doi</span> <span class="o">=</span> <span class="s2">&quot;10.1038/sdata.2016.18&quot;</span> <span class="c1">#https://doi.org/10.144534/rmh0000008&quot;</span>
<span class="n">publication</span><span class="o">.</span><span class="n">status</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;published&quot;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">publications</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">publication</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-the-isa-study-graph">
<h2>2. Create the ISA Study graph<a class="headerlink" href="#create-the-isa-study-graph" title="Permalink to this headline">Â¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To create the study graph that corresponds to the contents of the study table file (the s_*.txt file), we need</span>
<span class="c1"># to create a process sequence. To do this we use the Process class and attach it to the Study object&#39;s</span>
<span class="c1"># &#39;process_sequence&#39; list instance variable. Each process must be linked with a Protocol object that is attached to</span>
<span class="c1"># a Study object&#39;s &#39;protocols&#39; list instance variable. The sample collection Process object usually has as input</span>
<span class="c1"># a Source material and as output a Sample material.</span>

<span class="c1"># Here we create one Source material object and attach it to our study.</span>

<span class="n">source</span> <span class="o">=</span> <span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;source_material&#39;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

<span class="c1"># Then we create three Sample objects, with organism as Homo Sapiens, and attach them to the study. We use the utility function</span>
<span class="c1"># batch_create_material() to clone a prototype material object. The function automatiaclly appends</span>
<span class="c1"># an index to the material name. In this case, three samples will be created, with the names</span>
<span class="c1"># &#39;sample_material-0&#39;, &#39;sample_material-1&#39; and &#39;sample_material-2&#39;.</span>

<span class="n">prototype_sample</span> <span class="o">=</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sample_material&#39;</span><span class="p">,</span> <span class="n">derives_from</span><span class="o">=</span><span class="p">(</span><span class="n">source</span><span class="p">,))</span>
<span class="n">ncbitaxon</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;NCBITaxon&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;NCBI Taxonomy&quot;</span><span class="p">)</span>
<span class="n">characteristic_organism</span> <span class="o">=</span> <span class="n">Characteristic</span><span class="p">(</span><span class="n">category</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Organism&quot;</span><span class="p">),</span>
                                 <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Homo Sapiens&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">ncbitaxon</span><span class="p">,</span>
                                                          <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.bioontology.org/ontology/NCBITAXON/9606&quot;</span><span class="p">))</span>
<span class="n">prototype_sample</span><span class="o">.</span><span class="n">characteristics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">characteristic_organism</span><span class="p">)</span>

<span class="n">study</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">batch_create_materials</span><span class="p">(</span><span class="n">prototype_sample</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># creates a batch of 3 samples</span>
</pre></div>
</div>
</div>
</div>
<p>Now we create a single Protocol object that represents our sample collection protocol, and attach it to the
study object. Protocols must be declared before we describe Processes, as a processing event of some sort
must execute some defined protocol. In the case of the class model, Protocols should therefore be declared
before Processes in order for the Process to be linked to one.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample_collection_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">,</span>
                                      <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample_collection_protocol</span><span class="p">)</span>
<span class="n">sample_collection_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sample_collection_protocol</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we link our materials to the Process. In this particular case, we are describing a sample collection
process that takes one source material, and produces three different samples.</p>
<p>(source_material)-&gt;(sample collection)-&gt;[(sample_material-0), (sample_material-1), (sample_material-2)]</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">src</span> <span class="ow">in</span> <span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="p">:</span>
    <span class="n">sample_collection_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src</span><span class="p">)</span>
<span class="k">for</span> <span class="n">sam</span> <span class="ow">in</span> <span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">:</span>
    <span class="n">sample_collection_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sam</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, attach the finished Process object to the study process_sequence. This can be done many times to describe multiple sample collection events.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">study</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample_collection_process</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="building-an-assay-object-and-attach-two-protocols-extraction-and-sequencing">
<h2>3. Building an Assay object and attach two protocols, extraction and sequencing.<a class="headerlink" href="#building-an-assay-object-and-attach-two-protocols-extraction-and-sequencing" title="Permalink to this headline">Â¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Next, we build n Assay object and attach two protocols, extraction and sequencing.</span>

<span class="n">assay</span> <span class="o">=</span> <span class="n">Assay</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;a_assay.txt&quot;</span><span class="p">)</span>

<span class="n">extraction_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;extraction&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;material extraction&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_protocol</span><span class="p">)</span>

<span class="n">sequencing_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sequencing&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;material sequencing&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_protocol</span><span class="p">)</span>

<span class="c1"># To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an</span>
<span class="c1"># extraction process and a sequencing process. The extraction process takes as input a sample material, and produces</span>
<span class="c1"># an extract material. The sequencing process takes the extract material and produces a data file. This will</span>
<span class="c1"># produce three graphs, from sample material through to data, as follows:</span>
<span class="c1">#</span>
<span class="c1"># (sample_material-0)-&gt;(extraction)-&gt;(extract-0)-&gt;(sequencing)-&gt;(sequenced-data-0)</span>
<span class="c1"># (sample_material-1)-&gt;(extraction)-&gt;(extract-1)-&gt;(sequencing)-&gt;(sequenced-data-1)</span>
<span class="c1"># (sample_material-2)-&gt;(extraction)-&gt;(extract-2)-&gt;(sequencing)-&gt;(sequenced-data-2)</span>
<span class="c1">#</span>
<span class="c1"># Note that the extraction processes and sequencing processes are distinctly separate instances, where the three</span>
<span class="c1"># graphs are NOT interconnected.</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>

    <span class="c1"># create an extraction process that executes the extraction protocol</span>

    <span class="n">extraction_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">extraction_protocol</span><span class="p">)</span>

    <span class="c1"># extraction process takes as input a sample, and produces an extract material as output</span>

    <span class="n">extraction_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">material</span> <span class="o">=</span> <span class="n">Material</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;extract-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">material</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Extract Name&quot;</span>
    <span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>

    <span class="c1"># create a sequencing process that executes the sequencing protocol</span>

    <span class="n">sequencing_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sequencing_protocol</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;assay-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Sequencing process usually has an output data file</span>

    <span class="n">datafile</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sequenced-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Raw Data File&quot;</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>

    <span class="c1"># Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set</span>
    <span class="c1"># these links for you. It is found in the isatools.model package</span>

    <span class="n">plink</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">,</span> <span class="n">sequencing_process</span><span class="p">)</span>

    <span class="c1"># make sure the extract, data file, and the processes are attached to the assay</span>

    <span class="n">assay</span><span class="o">.</span><span class="n">data_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">other_material</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">measurement_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;genome sequencing&quot;</span><span class="p">)</span>
    <span class="n">assay</span><span class="o">.</span><span class="n">technology_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleotide sequencing&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We start by creation an ISA Assay Object and declaring the two protocols required by the workflow.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">assay_1</span> <span class="o">=</span> <span class="n">Assay</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;a_assay_1.txt&quot;</span><span class="p">)</span>

<span class="n">extraction_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;extraction&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;material extraction&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_protocol</span><span class="p">)</span>

<span class="n">sequencing_protocol_1</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sequencing_1&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid sequencing&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_protocol_1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an
extraction process and a sequencing process. The extraction process takes as input a sample material, and produces
an extract material. The sequencing process takes the extract material and produces a data file. This will
produce three graphs, from sample material through to data, as follows:</p>
<p><code class="docutils literal notranslate"><span class="pre">(sample_material-0)-&gt;(extraction)-&gt;(extract-0)-&gt;(sequencing)-&gt;(sequenced-data-0)</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">(sample_material-1)-&gt;(extraction)-&gt;(extract-1)-&gt;(sequencing)-&gt;(sequenced-data-1)</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">(sample_material-2)-&gt;(extraction)-&gt;(extract-2)-&gt;(sequencing)-&gt;(sequenced-data-2)</span></code></p>
<p>:warning:
Note that the extraction processes and sequencing processes are distinctly separate instances, where the three
graphs are NOT interconnected.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>

    <span class="c1"># create an extraction process that executes the extraction protocol</span>

    <span class="n">extraction_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">extraction_protocol</span><span class="p">)</span>

    <span class="c1"># extraction process takes as input a sample, and produces an extract material as output</span>

    <span class="n">extraction_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">material</span> <span class="o">=</span> <span class="n">Material</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;extract-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">material</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Extract Name&quot;</span>
    <span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>

    <span class="c1"># create a sequencing process that executes the sequencing protocol</span>

    <span class="n">sequencing_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sequencing_protocol_1</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;assay-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Sequencing process usually has an output data file</span>

    <span class="n">datafile</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sequenced-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Raw Data File&quot;</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>

    <span class="c1"># Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set</span>
    <span class="c1"># these links for you. It is found in the isatools.model package</span>

    <span class="n">plink</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">,</span> <span class="n">sequencing_process</span><span class="p">)</span>

    <span class="c1"># make sure the extract, data file, and the processes are attached to the assay</span>

    <span class="n">assay_1</span><span class="o">.</span><span class="n">data_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">other_material</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">measurement_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;genome sequencing&quot;</span><span class="p">)</span>
    <span class="n">assay_1</span><span class="o">.</span><span class="n">technology_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleotide sequencing&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Next, we build n Assay objects and attach the protocols, extraction, sequencing and a data transformation step</span>

<span class="n">assay_2</span> <span class="o">=</span> <span class="n">Assay</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;a_assay_2.txt&quot;</span><span class="p">)</span>
<span class="n">extraction_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;nucleic acid extraction&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid extraction&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_protocol</span><span class="p">)</span>
<span class="n">sequencing_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sequencing_2&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleic acid sequencing&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_protocol</span><span class="p">)</span>
<span class="n">dt_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;analysis&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sequence analysis data transformation&quot;</span><span class="p">))</span>
<span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt_protocol</span><span class="p">)</span>

<span class="c1"># To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an</span>
<span class="c1"># extraction process and a sequencing process. The extraction process takes as input a sample material, and produces</span>
<span class="c1"># an extract material. The sequencing process takes the extract material and produces a data file. This will</span>
<span class="c1"># produce three graphs, from sample material through to data, as follows:</span>
<span class="c1">#</span>
<span class="c1"># (sample_material-0)-&gt;(extraction)-&gt;(extract-0)-&gt;(sequencing)-&gt;(sequenced-data-0)</span>
<span class="c1"># (sample_material-1)-&gt;(extraction)-&gt;(extract-1)-&gt;(sequencing)-&gt;(sequenced-data-1)</span>
<span class="c1"># (sample_material-2)-&gt;(extraction)-&gt;(extract-2)-&gt;(sequencing)-&gt;(sequenced-data-2)</span>
<span class="c1">#</span>
<span class="c1"># Note that the extraction processes and sequencing processes are distinctly separate instances, where the three</span>
<span class="c1"># graphs are NOT interconnected.</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>

    <span class="c1"># create an extraction process that executes the extraction protocol</span>

    <span class="n">extraction_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">extraction_protocol</span><span class="p">)</span>

    <span class="c1"># extraction process takes as input a sample, and produces an extract material as output</span>

    <span class="n">extraction_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">material</span> <span class="o">=</span> <span class="n">Material</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;extract-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">material</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Extract Name&quot;</span>
    <span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>

    <span class="c1"># create a sequencing process that executes the sequencing protocol</span>

    <span class="n">sequencing_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sequencing_protocol</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;assay-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Sequencing process usually has an output data file</span>

    <span class="n">datafile</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sequenced-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Raw Data File&quot;</span><span class="p">)</span>
    <span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    
    <span class="c1"># Data Analysis process</span>
    <span class="n">dt_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">dt_protocol</span><span class="p">)</span>
    <span class="n">dt_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;dt-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">dt_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="n">derived_file</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;derived-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Derived Data File&quot;</span><span class="p">)</span>
    <span class="n">dt_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">derived_file</span><span class="p">)</span>
    
    <span class="c1"># Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set</span>
    <span class="c1"># these links for you. It is found in the isatools.model package</span>

    <span class="n">plink</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">,</span> <span class="n">sequencing_process</span><span class="p">)</span>
    <span class="n">plink</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">,</span> <span class="n">dt_process</span><span class="p">)</span>

    <span class="c1"># make sure the extract, data file, and the processes are attached to the assay</span>

    <span class="n">assay_2</span><span class="o">.</span><span class="n">data_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">other_material</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt_process</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">measurement_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;transcription profiling&quot;</span><span class="p">)</span>
    <span class="n">assay_2</span><span class="o">.</span><span class="n">technology_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleotide sequencing&quot;</span><span class="p">)</span>


<span class="nb">print</span><span class="p">(</span><span class="n">assay_2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Assay(
    measurement_type=transcription profiling
    technology_type=nucleotide sequencing
    technology_platform=
    filename=a_assay_2.txt
    data_files=3 DataFile objects
    samples=3 Sample objects
    process_sequence=9 Process objects
    other_material=3 Material objects
    characteristic_categories=0 OntologyAnnots
    comments=0 Comment objects
    units=0 Unit objects
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># attach the assays to the study</span>

<span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">assay_1</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">assay_2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataframes</span> <span class="o">=</span> <span class="n">dump_tables_to_dataframes</span><span class="p">(</span><span class="n">investigation</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2021-07-21 17:40:39,386 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [0]
2021-07-21 17:40:39,386 [WARNING]: isatab.py(write_study_table_files:1194) &gt;&gt; [5, 2, 3, 4, 0]
2021-07-21 17:40:39,387 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[0, 5, 2], [0, 5, 3], [0, 5, 4]]
2021-07-21 17:40:39,426 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,426 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 6, 7, 8], [3, 10, 11, 12], [4, 14, 15, 16]]
2021-07-21 17:40:39,427 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 6, 7, 8], [3, 10, 11, 12], [4, 14, 15, 16]]
2021-07-21 17:40:39,435 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,436 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 18, 19, 20], [3, 22, 23, 24], [4, 26, 27, 28]]
2021-07-21 17:40:39,437 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 18, 19, 20], [3, 22, 23, 24], [4, 26, 27, 28]]
2021-07-21 17:40:39,446 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,447 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 30, 31, 32, 34], [3, 36, 37, 38, 40], [4, 42, 43, 44, 46]]
2021-07-21 17:40:39,447 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 30, 31, 32, 34], [3, 36, 37, 38, 40], [4, 42, 43, 44, 46]]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">dataframes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">display</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;s_study.txt&#39;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;a_assay_2.txt&#39;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;a_assay.txt&#39;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;a_assay_1.txt&#39;
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="inspecting-the-different-assay-tables-reveals-the-importance-of-using-the-right-strings">
<h2>Inspecting the different assay tables reveals the importance of using the right strings<a class="headerlink" href="#inspecting-the-different-assay-tables-reveals-the-importance-of-using-the-right-strings" title="Permalink to this headline">Â¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="n">dataframes</span><span class="p">[</span><span class="s1">&#39;a_assay.txt&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample Name</th>
      <th>Protocol REF</th>
      <th>Extract Name</th>
      <th>Protocol REF.1</th>
      <th>Raw Data File</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sample_material-0</td>
      <td>extraction</td>
      <td>extract-0</td>
      <td>sequencing</td>
      <td>sequenced-data-0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>sample_material-1</td>
      <td>extraction</td>
      <td>extract-1</td>
      <td>sequencing</td>
      <td>sequenced-data-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>sample_material-2</td>
      <td>extraction</td>
      <td>extract-2</td>
      <td>sequencing</td>
      <td>sequenced-data-2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="n">dataframes</span><span class="p">[</span><span class="s1">&#39;a_assay_1.txt&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample Name</th>
      <th>Protocol REF</th>
      <th>Extract Name</th>
      <th>Protocol REF.1</th>
      <th>Assay Name</th>
      <th>Raw Data File</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sample_material-0</td>
      <td>extraction</td>
      <td>extract-0</td>
      <td>sequencing_1</td>
      <td>assay-name-0</td>
      <td>sequenced-data-0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>sample_material-1</td>
      <td>extraction</td>
      <td>extract-1</td>
      <td>sequencing_1</td>
      <td>assay-name-1</td>
      <td>sequenced-data-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>sample_material-2</td>
      <td>extraction</td>
      <td>extract-2</td>
      <td>sequencing_1</td>
      <td>assay-name-2</td>
      <td>sequenced-data-2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="n">dataframes</span><span class="p">[</span><span class="s1">&#39;a_assay_2.txt&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample Name</th>
      <th>Protocol REF</th>
      <th>Extract Name</th>
      <th>Protocol REF.1</th>
      <th>Assay Name</th>
      <th>Raw Data File</th>
      <th>Protocol REF.2</th>
      <th>Data Transformation Name</th>
      <th>Derived Data File</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sample_material-0</td>
      <td>nucleic acid extraction</td>
      <td>extract-0</td>
      <td>sequencing_2</td>
      <td>assay-name-0</td>
      <td>sequenced-data-0</td>
      <td>analysis</td>
      <td>dt-name-0</td>
      <td>derived-data-0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>sample_material-1</td>
      <td>nucleic acid extraction</td>
      <td>extract-1</td>
      <td>sequencing_2</td>
      <td>assay-name-1</td>
      <td>sequenced-data-1</td>
      <td>analysis</td>
      <td>dt-name-1</td>
      <td>derived-data-1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>sample_material-2</td>
      <td>nucleic acid extraction</td>
      <td>extract-2</td>
      <td>sequencing_2</td>
      <td>assay-name-2</td>
      <td>sequenced-data-2</td>
      <td>analysis</td>
      <td>dt-name-2</td>
      <td>derived-data-2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">isatab</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">investigation</span><span class="p">,</span><span class="s1">&#39;./notebook-output/isa-protocol-type-assay/&#39;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2021-07-21 17:40:39,603 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [0]
2021-07-21 17:40:39,604 [WARNING]: isatab.py(write_study_table_files:1194) &gt;&gt; [5, 2, 3, 4, 0]
2021-07-21 17:40:39,604 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[0, 5, 2], [0, 5, 3], [0, 5, 4]]
2021-07-21 17:40:39,624 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,625 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 6, 7, 8], [3, 10, 11, 12], [4, 14, 15, 16]]
2021-07-21 17:40:39,625 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 6, 7, 8], [3, 10, 11, 12], [4, 14, 15, 16]]
2021-07-21 17:40:39,633 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,634 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 18, 19, 20], [3, 22, 23, 24], [4, 26, 27, 28]]
2021-07-21 17:40:39,634 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 18, 19, 20], [3, 22, 23, 24], [4, 26, 27, 28]]
2021-07-21 17:40:39,643 [INFO]: isatab.py(_all_end_to_end_paths:1131) &gt;&gt; [2, 3, 4]
2021-07-21 17:40:39,644 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 30, 31, 32, 34], [3, 36, 37, 38, 40], [4, 42, 43, 44, 46]]
2021-07-21 17:40:39,644 [INFO]: isatab.py(_longest_path_and_attrs:1091) &gt;&gt; [[2, 30, 31, 32, 34], [3, 36, 37, 38, 40], [4, 42, 43, 44, 46]]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>isatools.model.Investigation(identifier=&#39;&#39;, filename=&#39;&#39;, title=&#39;&#39;, submission_date=&#39;&#39;, public_release_date=&#39;&#39;, ontology_source_references=[isatools.model.OntologySource(name=&#39;OBI&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;Ontology for Biomedical Investigations&#39;, comments=[])], publications=[], contacts=[], studies=[isatools.model.Study(filename=&#39;s_study.txt&#39;, identifier=&#39;s1&#39;, title=&#39;My ISA Study&#39;, description=&#39;Like with the Investigation, we could use the class constructor to set some default values, but have chosen to demonstrate in this example the use of instance variables to set initial values.&#39;, submission_date=&#39;2016-11-03&#39;, public_release_date=&#39;2016-11-03&#39;, contacts=[isatools.model.Person(last_name=&#39;Robertson&#39;, first_name=&#39;Alice&#39;, mid_initials=&#39;&#39;, email=&#39;&#39;, phone=&#39;&#39;, fax=&#39;&#39;, address=&#39;&#39;, affiliation=&#39;University of Life&#39;, roles=[isatools.model.OntologyAnnotation(term=&#39;submitter&#39;, term_source=None, term_accession=&#39;&#39;, comments=[])], comments=[])], design_descriptors=[isatools.model.OntologyAnnotation(term=&#39;intervention design&#39;, term_source=isatools.model.OntologySource(name=&#39;OBI&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;Ontology for Biomedical Investigations&#39;, comments=[]), term_accession=&#39;http://purl.obolibrary.org/obo/OBI_0000115&#39;, comments=[])], publications=[isatools.model.Publication(pubmed_id=&#39;12345678&#39;, doi=&#39;10.1038/sdata.2016.18&#39;, author_list=&#39;A. Robertson, B. Robertson&#39;, title=&#39;Experiments with Elephants&#39;, status=isatools.model.OntologyAnnotation(term=&#39;published&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), comments=[])], factors=[], protocols=[isatools.model.Protocol(name=&#39;sample collection&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;sample collection&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;extraction&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;material extraction&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;sequencing&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;material sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;extraction&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;material extraction&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;sequencing_1&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;nucleic acid sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;nucleic acid extraction&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;nucleic acid extraction&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;sequencing_2&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;nucleic acid sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[]), isatools.model.Protocol(name=&#39;analysis&#39;, protocol_type=isatools.model.OntologyAnnotation(term=&#39;sequence analysis data transformation&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), uri=&#39;&#39;, version=&#39;&#39;, parameters=[], components=[], comments=[])], assays=[isatools.model.Assay(measurement_type=isatools.model.OntologyAnnotation(term=&#39;genome sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_type=isatools.model.OntologyAnnotation(term=&#39;nucleotide sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_platform=&#39;&#39;, filename=&#39;a_assay.txt&#39;, data_files=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], samples=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200d82b0&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-0&quot;, executes_protocol=Protocol(
    name=sequencing
    protocol_type=material sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200d82b0&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200d8160&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-1&quot;, executes_protocol=Protocol(
    name=sequencing
    protocol_type=material sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200d8160&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x10a94cd90&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-2&quot;, executes_protocol=Protocol(
    name=sequencing
    protocol_type=material sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x10a94cd90&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])])], other_material=[&lt;isatools.model.Material object at 0x1200d82b0&gt;, &lt;isatools.model.Material object at 0x1200d8160&gt;, &lt;isatools.model.Material object at 0x10a94cd90&gt;], characteristic_categories=[], comments=[], units=[]), isatools.model.Assay(measurement_type=isatools.model.OntologyAnnotation(term=&#39;genome sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_type=isatools.model.OntologyAnnotation(term=&#39;nucleotide sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_platform=&#39;&#39;, filename=&#39;a_assay_1.txt&#39;, data_files=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], samples=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200b3d90&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-0&quot;, executes_protocol=Protocol(
    name=sequencing_1
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200b3d90&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200b3ee0&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-1&quot;, executes_protocol=Protocol(
    name=sequencing_1
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200b3ee0&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=extraction
    protocol_type=material extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200b3a00&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-2&quot;, executes_protocol=Protocol(
    name=sequencing_1
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200b3a00&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])])], other_material=[&lt;isatools.model.Material object at 0x1200b3d90&gt;, &lt;isatools.model.Material object at 0x1200b3ee0&gt;, &lt;isatools.model.Material object at 0x1200b3a00&gt;], characteristic_categories=[], comments=[], units=[]), isatools.model.Assay(measurement_type=isatools.model.OntologyAnnotation(term=&#39;transcription profiling&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_type=isatools.model.OntologyAnnotation(term=&#39;nucleotide sequencing&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), technology_platform=&#39;&#39;, filename=&#39;a_assay_2.txt&#39;, data_files=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[]), isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], samples=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=nucleic acid extraction
    protocol_type=nucleic acid extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200d7700&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-0&quot;, executes_protocol=Protocol(
    name=sequencing_2
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200d7700&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;dt-name-0&quot;, executes_protocol=Protocol(
    name=analysis
    protocol_type=sequence analysis data transformation
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.DataFile(filename=&#39;sequenced-data-0&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename=&#39;derived-data-0&#39;, label=&#39;Derived Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=nucleic acid extraction
    protocol_type=nucleic acid extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200d7dc0&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-1&quot;, executes_protocol=Protocol(
    name=sequencing_2
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200d7dc0&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;dt-name-1&quot;, executes_protocol=Protocol(
    name=analysis
    protocol_type=sequence analysis data transformation
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.DataFile(filename=&#39;sequenced-data-1&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename=&#39;derived-data-1&#39;, label=&#39;Derived Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=nucleic acid extraction
    protocol_type=nucleic acid extraction
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], outputs=[&lt;isatools.model.Material object at 0x1200d7340&gt;]), isatools.model.Process(id=&quot;&quot;. name=&quot;assay-name-2&quot;, executes_protocol=Protocol(
    name=sequencing_2
    protocol_type=nucleic acid sequencing
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[&lt;isatools.model.Material object at 0x1200d7340&gt;], outputs=[isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])]), isatools.model.Process(id=&quot;&quot;. name=&quot;dt-name-2&quot;, executes_protocol=Protocol(
    name=analysis
    protocol_type=sequence analysis data transformation
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.DataFile(filename=&#39;sequenced-data-2&#39;, label=&#39;Raw Data File&#39;, generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename=&#39;derived-data-2&#39;, label=&#39;Derived Data File&#39;, generated_from=[], comments=[])])], other_material=[&lt;isatools.model.Material object at 0x1200d7700&gt;, &lt;isatools.model.Material object at 0x1200d7dc0&gt;, &lt;isatools.model.Material object at 0x1200d7340&gt;], characteristic_categories=[], comments=[], units=[])], sources=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], samples=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=&quot;&quot;. name=&quot;None&quot;, executes_protocol=Protocol(
    name=sample collection
    protocol_type=sample collection
    uri=
    version=
    parameters=0 ProtocolParameter objects
    components=0 OntologyAnnotation objects
    comments=0 Comment objects
), date=&quot;None&quot;, performer=&quot;None&quot;, inputs=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], outputs=[isatools.model.Sample(name=&#39;sample_material-0&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-1&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name=&#39;sample_material-2&#39;, characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term=&#39;Organism&#39;, term_source=None, term_accession=&#39;&#39;, comments=[]), value=isatools.model.OntologyAnnotation(term=&#39;Homo Sapiens&#39;, term_source=isatools.model.OntologySource(name=&#39;NCBITaxon&#39;, file=&#39;&#39;, version=&#39;&#39;, description=&#39;NCBI Taxonomy&#39;, comments=[]), term_accession=&#39;http://purl.bioontology.org/ontology/NCBITAXON/9606&#39;, comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name=&#39;source_material&#39;, characteristics=[], comments=[])], comments=[])])], other_material=[], characteristic_categories=[], comments=[], units=[])], comments=[])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">my_json_report</span> <span class="o">=</span> <span class="n">isatab</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;./notebook-output/isa-protocol-type-assay/&#39;</span><span class="p">,</span> <span class="s1">&#39;i_investigation.txt&#39;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2021-07-21 17:40:39,664 [INFO]: isatab.py(validate:4206) &gt;&gt; Loading... ./notebook-output/isa-protocol-type-assay/i_investigation.txt
2021-07-21 17:40:39,778 [INFO]: isatab.py(validate:4208) &gt;&gt; Running prechecks...
2021-07-21 17:40:39,834 [WARNING]: isatab.py(check_doi:2176) &gt;&gt; (W) DOI 10.1038/sdata.2016.18 does not conform to DOI format
2021-07-21 17:40:39,835 [INFO]: isatab.py(validate:4229) &gt;&gt; Finished prechecks...
2021-07-21 17:40:39,836 [INFO]: isatab.py(validate:4230) &gt;&gt; Loading configurations found in /Users/philippe/.pyenv/versions/3.9.0/envs/isa-api-py39/lib/python3.9/site-packages/isatools/resources/config/xml
2021-07-21 17:40:39,857 [INFO]: isatab.py(validate:4235) &gt;&gt; Using configurations found in /Users/philippe/.pyenv/versions/3.9.0/envs/isa-api-py39/lib/python3.9/site-packages/isatools/resources/config/xml
2021-07-21 17:40:39,857 [INFO]: isatab.py(validate:4237) &gt;&gt; Checking investigation file against configuration...
2021-07-21 17:40:39,859 [WARNING]: isatab.py(check_section_against_required_fields_one_value:3363) &gt;&gt; (W) A property value in Study Person Mid Initials of investigation file at column 1 is required
2021-07-21 17:40:39,860 [INFO]: isatab.py(validate:4240) &gt;&gt; Finished checking investigation file
2021-07-21 17:40:39,860 [INFO]: isatab.py(validate:4259) &gt;&gt; Loading... s_study.txt
2021-07-21 17:40:39,863 [INFO]: isatab.py(validate:4265) &gt;&gt; Validating s_study.txt against default study table configuration
2021-07-21 17:40:39,864 [INFO]: isatab.py(validate:4267) &gt;&gt; Checking Factor Value presence...
2021-07-21 17:40:39,864 [INFO]: isatab.py(validate:4270) &gt;&gt; Checking required fields...
2021-07-21 17:40:39,865 [INFO]: isatab.py(validate:4273) &gt;&gt; Checking generic fields...
2021-07-21 17:40:39,866 [INFO]: isatab.py(validate:4281) &gt;&gt; Checking unit fields...
2021-07-21 17:40:39,866 [INFO]: isatab.py(validate:4288) &gt;&gt; Checking protocol fields...
2021-07-21 17:40:39,867 [INFO]: isatab.py(validate:4298) &gt;&gt; Checking ontology fields...
2021-07-21 17:40:39,868 [INFO]: isatab.py(validate:4308) &gt;&gt; Checking study group size...
2021-07-21 17:40:39,868 [INFO]: isatab.py(validate:4312) &gt;&gt; Finished validation on s_study.txt
2021-07-21 17:40:39,869 [INFO]: isatab.py(validate:4349) &gt;&gt; Loading... a_assay.txt
2021-07-21 17:40:39,872 [INFO]: isatab.py(validate:4357) &gt;&gt; Validating a_assay.txt against assay table configuration (genome sequencing, nucleotide sequencing)...
2021-07-21 17:40:39,872 [INFO]: isatab.py(validate:4363) &gt;&gt; Checking Factor Value presence...
2021-07-21 17:40:39,873 [INFO]: isatab.py(validate:4367) &gt;&gt; Checking required fields...
2021-07-21 17:40:39,873 [WARNING]: isatab.py(check_required_fields:3687) &gt;&gt; (W) Required field &#39;Assay Name&#39; not found in the file &#39;a_assay.txt&#39;
2021-07-21 17:40:39,874 [INFO]: isatab.py(validate:4370) &gt;&gt; Checking generic fields...
2021-07-21 17:40:39,875 [INFO]: isatab.py(validate:4381) &gt;&gt; Checking unit fields...
2021-07-21 17:40:39,876 [INFO]: isatab.py(validate:4391) &gt;&gt; Checking protocol fields...
2021-07-21 17:40:39,877 [WARNING]: isatab.py(check_protocol_fields:4018) &gt;&gt; (W) Protocol(s) of type [&#39;nucleic acid extraction&#39;] defined in the ISA-configuration expected as a between &#39;Sample Name&#39; and &#39;Extract Name&#39; but has not been found, in the file &#39;a_assay.txt&#39;
2021-07-21 17:40:39,877 [WARNING]: isatab.py(check_protocol_fields:4018) &gt;&gt; (W) Protocol(s) of type [&#39;library construction&#39;, &#39;nucleic acid sequencing&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Raw Data File&#39; but has not been found, in the file &#39;a_assay.txt&#39;
2021-07-21 17:40:39,878 [WARNING]: isatab.py(validate:4396) &gt;&gt; (W) There are some protocol inconsistencies in a_assay.txt against (&#39;genome sequencing&#39;, &#39;nucleotide sequencing&#39;) configuration
2021-07-21 17:40:39,878 [INFO]: isatab.py(validate:4403) &gt;&gt; Checking ontology fields...
2021-07-21 17:40:39,879 [INFO]: isatab.py(validate:4416) &gt;&gt; Checking study group size...
2021-07-21 17:40:39,879 [INFO]: isatab.py(validate:4420) &gt;&gt; Finished validation on a_assay.txt
2021-07-21 17:40:39,880 [INFO]: isatab.py(validate:4426) &gt;&gt; Checking consistencies between study sample table and assay tables...
2021-07-21 17:40:39,880 [INFO]: isatab.py(validate:4431) &gt;&gt; Finished checking study sample table against assay tables...
2021-07-21 17:40:39,881 [INFO]: isatab.py(validate:4349) &gt;&gt; Loading... a_assay_1.txt
2021-07-21 17:40:39,884 [INFO]: isatab.py(validate:4357) &gt;&gt; Validating a_assay_1.txt against assay table configuration (genome sequencing, nucleotide sequencing)...
2021-07-21 17:40:39,884 [INFO]: isatab.py(validate:4363) &gt;&gt; Checking Factor Value presence...
2021-07-21 17:40:39,885 [INFO]: isatab.py(validate:4367) &gt;&gt; Checking required fields...
2021-07-21 17:40:39,885 [INFO]: isatab.py(validate:4370) &gt;&gt; Checking generic fields...
2021-07-21 17:40:39,887 [INFO]: isatab.py(validate:4381) &gt;&gt; Checking unit fields...
2021-07-21 17:40:39,888 [INFO]: isatab.py(validate:4391) &gt;&gt; Checking protocol fields...
2021-07-21 17:40:39,889 [WARNING]: isatab.py(check_protocol_fields:4018) &gt;&gt; (W) Protocol(s) of type [&#39;nucleic acid extraction&#39;] defined in the ISA-configuration expected as a between &#39;Sample Name&#39; and &#39;Extract Name&#39; but has not been found, in the file &#39;a_assay_1.txt&#39;
2021-07-21 17:40:39,889 [WARNING]: isatab.py(check_protocol_fields:4018) &gt;&gt; (W) Protocol(s) of type [&#39;library construction&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Assay Name&#39; but has not been found, in the file &#39;a_assay_1.txt&#39;
2021-07-21 17:40:39,890 [WARNING]: isatab.py(validate:4396) &gt;&gt; (W) There are some protocol inconsistencies in a_assay_1.txt against (&#39;genome sequencing&#39;, &#39;nucleotide sequencing&#39;) configuration
2021-07-21 17:40:39,890 [INFO]: isatab.py(validate:4403) &gt;&gt; Checking ontology fields...
2021-07-21 17:40:39,891 [INFO]: isatab.py(validate:4416) &gt;&gt; Checking study group size...
2021-07-21 17:40:39,891 [INFO]: isatab.py(validate:4420) &gt;&gt; Finished validation on a_assay_1.txt
2021-07-21 17:40:39,892 [INFO]: isatab.py(validate:4426) &gt;&gt; Checking consistencies between study sample table and assay tables...
2021-07-21 17:40:39,892 [INFO]: isatab.py(validate:4431) &gt;&gt; Finished checking study sample table against assay tables...
2021-07-21 17:40:39,893 [INFO]: isatab.py(validate:4349) &gt;&gt; Loading... a_assay_2.txt
2021-07-21 17:40:39,896 [INFO]: isatab.py(validate:4357) &gt;&gt; Validating a_assay_2.txt against assay table configuration (transcription profiling, nucleotide sequencing)...
2021-07-21 17:40:39,897 [INFO]: isatab.py(validate:4363) &gt;&gt; Checking Factor Value presence...
2021-07-21 17:40:39,897 [INFO]: isatab.py(validate:4367) &gt;&gt; Checking required fields...
2021-07-21 17:40:39,897 [WARNING]: isatab.py(check_required_fields:3687) &gt;&gt; (W) Required field &#39;Parameter Value[library layout]&#39; not found in the file &#39;a_assay_2.txt&#39;
2021-07-21 17:40:39,898 [INFO]: isatab.py(validate:4370) &gt;&gt; Checking generic fields...
2021-07-21 17:40:39,900 [INFO]: isatab.py(validate:4381) &gt;&gt; Checking unit fields...
2021-07-21 17:40:39,901 [INFO]: isatab.py(validate:4391) &gt;&gt; Checking protocol fields...
2021-07-21 17:40:39,902 [WARNING]: isatab.py(check_protocol_fields:4018) &gt;&gt; (W) Protocol(s) of type [&#39;library construction&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Assay Name&#39; but has not been found, in the file &#39;a_assay_2.txt&#39;
2021-07-21 17:40:39,903 [WARNING]: isatab.py(validate:4396) &gt;&gt; (W) There are some protocol inconsistencies in a_assay_2.txt against (&#39;transcription profiling&#39;, &#39;nucleotide sequencing&#39;) configuration
2021-07-21 17:40:39,903 [INFO]: isatab.py(validate:4403) &gt;&gt; Checking ontology fields...
2021-07-21 17:40:39,904 [INFO]: isatab.py(validate:4416) &gt;&gt; Checking study group size...
2021-07-21 17:40:39,904 [INFO]: isatab.py(validate:4420) &gt;&gt; Finished validation on a_assay_2.txt
2021-07-21 17:40:39,905 [INFO]: isatab.py(validate:4426) &gt;&gt; Checking consistencies between study sample table and assay tables...
2021-07-21 17:40:39,905 [INFO]: isatab.py(validate:4431) &gt;&gt; Finished checking study sample table against assay tables...
2021-07-21 17:40:40,141 [INFO]: utils.py(detect_isatab_process_pooling:76) &gt;&gt; Checking s_study.txt
2021-07-21 17:40:40,142 [INFO]: utils.py(detect_isatab_process_pooling:85) &gt;&gt; Checking a_assay.txt
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2021-07-21 17:40:40,142 [INFO]: utils.py(detect_isatab_process_pooling:85) &gt;&gt; Checking a_assay_1.txt
2021-07-21 17:40:40,143 [INFO]: utils.py(detect_isatab_process_pooling:85) &gt;&gt; Checking a_assay_2.txt
2021-07-21 17:40:40,143 [INFO]: utils.py(detect_graph_process_pooling:57) &gt;&gt; Possible process pooling detected on:  analysis
2021-07-21 17:40:40,144 [INFO]: utils.py(detect_graph_process_pooling:57) &gt;&gt; Possible process pooling detected on:  analysis
2021-07-21 17:40:40,145 [INFO]: utils.py(detect_graph_process_pooling:57) &gt;&gt; Possible process pooling detected on:  analysis
2021-07-21 17:40:40,145 [INFO]: isatab.py(validate:4444) &gt;&gt; Finished validation...
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">my_json_report</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;errors&#39;: [], &#39;warnings&#39;: [{&#39;message&#39;: &#39;DOI is not valid format&#39;, &#39;supplemental&#39;: &#39;Found 10.1038/sdata.2016.18 in DOI field&#39;, &#39;code&#39;: 3002}, {&#39;message&#39;: &#39;A required property is missing&#39;, &#39;supplemental&#39;: &#39;A property value in Study Person Mid Initials of investigation file at column 1 is required&#39;, &#39;code&#39;: 4003}, {&#39;message&#39;: &#39;A required column in assay table is not present&#39;, &#39;supplemental&#39;: &quot;Required field &#39;Assay Name&#39; not found in the file &#39;a_assay.txt&#39;&quot;, &#39;code&#39;: 4010}, {&#39;message&#39;: &#39;Missing Protocol declaration&#39;, &#39;supplemental&#39;: &quot;Protocol(s) of type [&#39;nucleic acid extraction&#39;] defined in the ISA-configuration expected as a between &#39;Sample Name&#39; and &#39;Extract Name&#39; but has not been found, in the file &#39;a_assay.txt&#39;&quot;, &#39;code&#39;: 1007}, {&#39;message&#39;: &#39;Missing Protocol declaration&#39;, &#39;supplemental&#39;: &quot;Protocol(s) of type [&#39;library construction&#39;, &#39;nucleic acid sequencing&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Raw Data File&#39; but has not been found, in the file &#39;a_assay.txt&#39;&quot;, &#39;code&#39;: 1007}, {&#39;message&#39;: &#39;Missing Protocol declaration&#39;, &#39;supplemental&#39;: &quot;Protocol(s) of type [&#39;nucleic acid extraction&#39;] defined in the ISA-configuration expected as a between &#39;Sample Name&#39; and &#39;Extract Name&#39; but has not been found, in the file &#39;a_assay_1.txt&#39;&quot;, &#39;code&#39;: 1007}, {&#39;message&#39;: &#39;Missing Protocol declaration&#39;, &#39;supplemental&#39;: &quot;Protocol(s) of type [&#39;library construction&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Assay Name&#39; but has not been found, in the file &#39;a_assay_1.txt&#39;&quot;, &#39;code&#39;: 1007}, {&#39;message&#39;: &#39;A required column in assay table is not present&#39;, &#39;supplemental&#39;: &quot;Required field &#39;Parameter Value[library layout]&#39; not found in the file &#39;a_assay_2.txt&#39;&quot;, &#39;code&#39;: 4010}, {&#39;message&#39;: &#39;Missing Protocol declaration&#39;, &#39;supplemental&#39;: &quot;Protocol(s) of type [&#39;library construction&#39;] defined in the ISA-configuration expected as a between &#39;Extract Name&#39; and &#39;Assay Name&#39; but has not been found, in the file &#39;a_assay_2.txt&#39;&quot;, &#39;code&#39;: 1007}], &#39;info&#39;: [{&#39;message&#39;: &#39;Found -1 study groups in s_study.txt&#39;, &#39;supplemental&#39;: &#39;Found -1 study groups in s_study.txt&#39;, &#39;code&#39;: 5001}, {&#39;message&#39;: &#39;Found -1 study groups in a_assay.txt&#39;, &#39;supplemental&#39;: &#39;Found -1 study groups in a_assay.txt&#39;, &#39;code&#39;: 5001}, {&#39;message&#39;: &#39;Found -1 study groups in a_assay_1.txt&#39;, &#39;supplemental&#39;: &#39;Found -1 study groups in a_assay_1.txt&#39;, &#39;code&#39;: 5001}, {&#39;message&#39;: &#39;Found -1 study groups in a_assay_2.txt&#39;, &#39;supplemental&#39;: &#39;Found -1 study groups in a_assay_2.txt&#39;, &#39;code&#39;: 5001}], &#39;validation_finished&#39;: True}
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="about-this-notebook">
<h2>About this notebook<a class="headerlink" href="#about-this-notebook" title="Permalink to this headline">Â¶</a></h2>
<ul class="simple">
<li><p>authors: <a class="reference external" href="mailto:philippe&#46;rocca-serra&#37;&#52;&#48;oerc&#46;ox&#46;ac&#46;uk">philippe<span>&#46;</span>rocca-serra<span>&#64;</span>oerc<span>&#46;</span>ox<span>&#46;</span>ac<span>&#46;</span>uk</a>, <a class="reference external" href="mailto:massimiliano&#46;izzo&#37;&#52;&#48;oerc&#46;ox&#46;ac&#46;uk">massimiliano<span>&#46;</span>izzo<span>&#64;</span>oerc<span>&#46;</span>ox<span>&#46;</span>ac<span>&#46;</span>uk</a></p></li>
<li><p>license: CC-BY 4.0</p></li>
<li><p>support: <a class="reference external" href="mailto:isatools&#37;&#52;&#48;googlegroups&#46;com">isatools<span>&#64;</span>googlegroups<span>&#46;</span>com</a></p></li>
<li><p>issue tracker: <a class="reference external" href="https://github.com/ISA-tools/isa-api/issues">https://github.com/ISA-tools/isa-api/issues</a></p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./content/notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="programmatic-rebuild-of-BII-S-3-with-ISA-API.html" title="previous page">Programmatic recreation of BII-S-3 with the ISA-API</a>
    <a class='right-next' id="next-link" href="create-a-simple-ISA-descriptor.html" title="next page">Create a simple ISA descriptor</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The ISA Team<br/>
        
            &copy; Copyright 2021 (CC-BY 4.0).<br/>
          <div class="extra_footer">
            <script>mermaid.init();</script> <div class="row_footer"> This page was created thanks to work by <a href="https://github.com/jupyter/jupyter-book/graphs/contributors">The Jupyter Book Community</a></div> <div class="row_footer"> This project is supported financially by the <a href="https://wellcome.org/">Welcome Trust</a>, Elixir-UK and the European Union</div>
<div class="row_footer"> <div class="column" style="center"><a href="https://wellcome.org/"><img src="../_static/images/logo/wellcome-logo-black.png" height="25" alt="Wellcome" style="float:center""></a></div> <div class="column"><a href="https://www.elixir-uk.org.eu/"><img src="../_static/images/logo/elixir-uk_transparent.png" height="40" alt="IMI logo" style="float:center""></a></div> <div class="column"><a href="https://https://ec.europa.eu/programmes/horizon2020/en"><img src="../_static/images/logo/eu-flag-recolored.png" height="25" alt="EFPIA logo" style="float:center""></a></div> </div>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>